{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7251a47c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_91708/3777615979.py:1: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display\n",
      "  from IPython.core.display import display, HTML\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cea41ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "import dynamo as dyn\n",
    "dyn.configuration.set_figure_params('dynamo', background='white')\n",
    "print(dyn.__path__)\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import sparse\n",
    "from scipy import stats\n",
    "from scipy.integrate import solve_ivp\n",
    "from dynamo.vectorfield import scVectorField\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import matplotlib.patches as mpatches\n",
    "import random\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "from sklearn.decomposition import PCA\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import pickle\n",
    "import os\n",
    "from scipy.cluster.hierarchy import dendrogram\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "from scipy.spatial.distance import euclidean\n",
    "from fastdtw import fastdtw\n",
    "from dynamo.vectorfield.utils import vector_transformation\n",
    "from dynamo.preprocessing.cell_cycle import get_cell_phase_genes\n",
    "import collections\n",
    "from scipy.sparse import issparse\n",
    "from collections import Counter\n",
    "from matplotlib import ticker\n",
    "from scipy.stats import ks_2samp\n",
    "import anndata as ad\n",
    "import umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "da2d6f4a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>package</th>\n",
       "      <th>umap-learn</th>\n",
       "      <th>tqdm</th>\n",
       "      <th>statsmodels</th>\n",
       "      <th>setuptools</th>\n",
       "      <th>seaborn</th>\n",
       "      <th>scipy</th>\n",
       "      <th>scikit-learn</th>\n",
       "      <th>python-igraph</th>\n",
       "      <th>pynndescent</th>\n",
       "      <th>pre-commit</th>\n",
       "      <th>pandas</th>\n",
       "      <th>openpyxl</th>\n",
       "      <th>numpy</th>\n",
       "      <th>numdifftools</th>\n",
       "      <th>numba</th>\n",
       "      <th>networkx</th>\n",
       "      <th>matplotlib</th>\n",
       "      <th>loompy</th>\n",
       "      <th>get-version</th>\n",
       "      <th>dynamo-release</th>\n",
       "      <th>cvxopt</th>\n",
       "      <th>colorcet</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>version</th>\n",
       "      <td>0.5.5</td>\n",
       "      <td>4.66.1</td>\n",
       "      <td>0.14.1</td>\n",
       "      <td>69.0.3</td>\n",
       "      <td>0.13.1</td>\n",
       "      <td>1.10.1</td>\n",
       "      <td>1.3.2</td>\n",
       "      <td>0.11.3</td>\n",
       "      <td>0.5.11</td>\n",
       "      <td>3.5.0</td>\n",
       "      <td>2.0.3</td>\n",
       "      <td>3.1.2</td>\n",
       "      <td>1.24.4</td>\n",
       "      <td>0.9.41</td>\n",
       "      <td>0.58.1</td>\n",
       "      <td>3.1</td>\n",
       "      <td>3.7.4</td>\n",
       "      <td>3.0.7</td>\n",
       "      <td>3.5.4</td>\n",
       "      <td>1.1.0</td>\n",
       "      <td>1.3.2</td>\n",
       "      <td>3.0.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "package umap-learn    tqdm statsmodels setuptools seaborn   scipy  \\\n",
       "version      0.5.5  4.66.1      0.14.1     69.0.3  0.13.1  1.10.1   \n",
       "\n",
       "package scikit-learn python-igraph pynndescent pre-commit pandas openpyxl  \\\n",
       "version        1.3.2        0.11.3      0.5.11      3.5.0  2.0.3    3.1.2   \n",
       "\n",
       "package   numpy numdifftools   numba networkx matplotlib loompy get-version  \\\n",
       "version  1.24.4       0.9.41  0.58.1      3.1      3.7.4  3.0.7       3.5.4   \n",
       "\n",
       "package dynamo-release cvxopt colorcet  \n",
       "version          1.1.0  1.3.2    3.0.1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dyn.get_all_dependencies_version()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aa4e0f41",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'plt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m#Define a nice colour map for gene expression\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mmatplotlib\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m colors\n\u001b[0;32m----> 3\u001b[0m colors2 \u001b[38;5;241m=\u001b[39m \u001b[43mplt\u001b[49m\u001b[38;5;241m.\u001b[39mcm\u001b[38;5;241m.\u001b[39mReds(np\u001b[38;5;241m.\u001b[39mlinspace(\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m128\u001b[39m))\n\u001b[1;32m      4\u001b[0m colors3 \u001b[38;5;241m=\u001b[39m plt\u001b[38;5;241m.\u001b[39mcm\u001b[38;5;241m.\u001b[39mGreys_r(np\u001b[38;5;241m.\u001b[39mlinspace(\u001b[38;5;241m0.7\u001b[39m,\u001b[38;5;241m0.8\u001b[39m,\u001b[38;5;241m20\u001b[39m))\n\u001b[1;32m      5\u001b[0m colorsComb \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mvstack([colors3, colors2])\n",
      "\u001b[0;31mNameError\u001b[0m: name 'plt' is not defined"
     ]
    }
   ],
   "source": [
    "#Define a nice colour map for gene expression\n",
    "from matplotlib import colors\n",
    "colors2 = plt.cm.Reds(np.linspace(0, 1, 128))\n",
    "colors3 = plt.cm.Greys_r(np.linspace(0.7,0.8,20))\n",
    "colorsComb = np.vstack([colors3, colors2])\n",
    "mymap = colors.LinearSegmentedColormap.from_list('my_colormap', colorsComb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81b37a80",
   "metadata": {},
   "outputs": [],
   "source": [
    "phases = ['G1.S', 'S', 'G2', 'G2.M', 'M.G1']\n",
    "colors = {'G1.S':'firebrick', 'S':'orange', 'G2':'yellowgreen', 'G2.M':'teal', 'M.G1':'royalblue'}\n",
    "doses = [0,12,25,50,100,200,400,800]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f227750",
   "metadata": {
    "heading_collapsed": true,
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Processing adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f96bd8a0",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Processing the data - individual UMAP reference\n",
    "def processAdata(adata, density = False):\n",
    "    dyn.tl.dynamics(adata, model='deterministic', cores=3)\n",
    "    dyn.tl.reduceDimension(adata, enforce=True)\n",
    "    dyn.tl.cell_wise_confidence(adata)\n",
    "    dyn.tl.cell_velocities(adata, method='pearson', enforce=True, other_kernels_dict={'transform': 'sqrt'}, correct_density = density)\n",
    "    dyn.tl.cell_velocities(adata, basis = 'pca', enforce = True, method='pearson', other_kernels_dict={'transform': 'sqrt'}, correct_density = density)\n",
    "    #dyn.tl.cell_velocities(adata, basis = 'DC', enforce = True, method='pearson', other_kernels_dict={'transform': 'sqrt'}, correct_density = density)\n",
    "    #visualize what the velocity looks like\n",
    "    dyn.vf.VectorField(adata, basis='pca', M=1000)\n",
    "    dyn.vf.VectorField(adata, basis='umap', M=1000)\n",
    "    dyn.vf.topography(adata, basis='umap')\n",
    "    dyn.vf.topography(adata, basis='pca')\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01e6537c",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# adata = not processed\n",
    "# processed_adata = processed adata (contains the UMAP and PCA coordinate reference)\n",
    "def processAdata_sameRef(adata, processed_adata, dosage, genes, keep_filtered = False, density = False):\n",
    "\n",
    "    adata_dose = adata[(dosage == adata.obs['dosage'])]\n",
    "    if genes is None:\n",
    "        dyn.pp.recipe_monocle(adata_dose, n_top_genes=1000, keep_filtered_genes=True)\n",
    "    else:\n",
    "        #adata_dose.var.index = adata_dose.var['var_names']\n",
    "        dyn.pp.recipe_monocle(adata_dose, genes_to_use=genes,  n_top_genes = len(genes), keep_filtered_genes=keep_filtered)\n",
    "    \n",
    "    X_umap = processed_adata.obsm['X_umap']\n",
    "    X_pca = processed_adata.obsm['X_pca']\n",
    "    \n",
    "    index = dosage == processed_adata.obs['dosage']\n",
    "    adata_dose.obsm['X_pca'] = X_pca[index]\n",
    "    \n",
    "    dyn.tl.dynamics(adata_dose, model='deterministic', cores=3)\n",
    "    \n",
    "    adata_dose.obsm['X_umap'] = X_umap[index]\n",
    "    \n",
    "    dyn.tl.reduceDimension(adata_dose)\n",
    "    \n",
    "    dyn.tl.cell_velocities(adata_dose, method='pearson', enforce=True, other_kernels_dict={'transform': 'sqrt'}, correct_density = density)\n",
    "    dyn.tl.cell_velocities(adata_dose, basis = 'pca', enforce = True, method='pearson', other_kernels_dict={'transform': 'sqrt'}, correct_density=density)\n",
    "    dyn.vf.VectorField(adata_dose, basis='pca', M=1000)\n",
    "    dyn.vf.VectorField(adata_dose, basis='umap', M=1000)\n",
    "    dyn.vf.topography(adata_dose, basis='pca')\n",
    "    dyn.vf.topography(adata_dose, basis='umap')\n",
    "    \n",
    "    return adata_dose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "8dcf99ce",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# adata = not processed\n",
    "# processed_adata = processed adata (contains the UMAP and PCA coordinate reference)\n",
    "def processAdata_sameRef2(adata, processed_adata, dosage, genes, dims=30, keep_filtered = False, density = False):\n",
    "    print('Dosage:', dosage)\n",
    "    adata_dose = adata[(dosage == adata.obs['dosage'])]\n",
    "    #adata.var.index = adata.var['var_names']\n",
    "    if genes is None:\n",
    "        dyn.pp.recipe_monocle(adata_dose, n_top_genes=1000, keep_filtered_genes=False)\n",
    "    else:\n",
    "        adata_dose.var.index = adata_dose.var['var_names']\n",
    "        dyn.pp.recipe_monocle(adata_dose, genes_to_use=genes,  n_top_genes = len(genes), keep_filtered_genes=keep_filtered)\n",
    "\n",
    "    X_umap = processed_adata.obsm['X_umap']\n",
    "    X_pca = processed_adata.obsm['X_pca'][:,0:dims]\n",
    "    \n",
    "    index = dosage == processed_adata.obs['dosage']\n",
    "\n",
    "    dyn.tl.dynamics(adata_dose, model='deterministic', cores=3)\n",
    "\n",
    "    adata_dose.obsm['X_pca'] = X_pca[index]\n",
    "    adata_dose.obsm['X_umap'] = X_umap[index]\n",
    "\n",
    "    adata_dose.uns['pca_fit'] = processed_adata.uns['pca_fit']\n",
    "    adata_dose.uns['umap_fit'] = processed_adata.uns['umap_fit']\n",
    "    \n",
    "    #dyn.tl.reduceDimension(adata_dose)\n",
    "    dyn.tl.neighbors(adata_dose)\n",
    "    dyn.tl.cell_velocities(adata_dose, method='pearson', enforce=True, other_kernels_dict={'transform': 'sqrt'}, correct_density = density)\n",
    "    dyn.tl.cell_velocities(adata_dose, basis = 'pca', enforce = True, method='pearson', other_kernels_dict={'transform': 'sqrt'}, correct_density=density)\n",
    "    dyn.vf.VectorField(adata_dose, basis='pca', M=1000)\n",
    "    dyn.vf.VectorField(adata_dose, basis='umap', M=1000)\n",
    "    dyn.vf.topography(adata_dose, basis='pca')\n",
    "    dyn.vf.topography(adata_dose, basis='umap')\n",
    "    \n",
    "    return adata_dose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "37b1f11f",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def process_ind(adata, dosage, genes, density):\n",
    "    adata_dose = adata[(dosage == adata.obs['dosage'])]\n",
    "    if genes is None:\n",
    "        dyn.pp.recipe_monocle(adata_dose, n_top_genes=1000, keep_filtered_genes=False)\n",
    "    else:\n",
    "        adata_dose.var.index = adata_dose.var['var_names']\n",
    "        dyn.pp.recipe_monocle(adata_dose, genes_to_use=genes, keep_filtered_genes=density)\n",
    "\n",
    "    dyn.tl.dynamics(adata_dose, model='deterministic', cores=3)\n",
    "    dyn.tl.reduceDimension(adata_dose)\n",
    "    dyn.tl.cell_wise_confidence(adata_dose)\n",
    "    dyn.tl.cell_velocities(adata_dose, method='pearson', enforce=True, other_kernels_dict={'transform': 'sqrt'}, correct_density = density)\n",
    "    dyn.tl.cell_velocities(adata_dose, basis = 'pca', enforce = True, method='pearson', other_kernels_dict={'transform': 'sqrt'}, correct_density = density)\n",
    "    dyn.vf.VectorField(adata_dose, basis='pca', M=1000)\n",
    "    dyn.vf.topography(adata_dose, basis='umap')\n",
    "    dyn.vf.topography(adata_dose, basis='pca')\n",
    "    return adata_dose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c304240d",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Obtains the acceleration and divergence rank for all the genes \n",
    "def get_rank_genes(adata):\n",
    "    valid_gene_list = adata.var['var_names'].astype(str)\n",
    "    dyn.vf.jacobian(adata, regulators=valid_gene_list.index, effectors=valid_gene_list.index)\n",
    "    dyn.vf.rank_acceleration_genes(adata)\n",
    "    \n",
    "    temp = np.column_stack((adata.uns['rank_acceleration'].astype(int), adata.var['var_names'][:,None]))\n",
    "    sorted_acceleration = temp[temp[:, 0].argsort()]\n",
    "    \n",
    "    dyn.vf.rank_divergence_genes(adata)\n",
    "    temp = np.column_stack((adata.uns['rank_div_gene_jacobian_pca'].astype(int), adata.var['var_names'][:,None]))\n",
    "    sorted_divergence = temp[temp[:, 0].argsort()]\n",
    "    \n",
    "    return sorted_acceleration, sorted_divergence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a27a0931",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def saveToTXT(toSave, fileName):\n",
    "    file = open(\"../generated-data/Gene Lists/\"+fileName, \"w\")\n",
    "    for i in range(toSave.shape[0]):\n",
    "        #file.write(toSave[i,1].astype(str)+\"\\n\")\n",
    "        file.write(toSave[i,0].astype(str)+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "90a1ddd6",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_jacobian_genes(adata, jkey=\"jacobian\", basis=\"pca\", regulators=None, effectors=None, **kwargs):\n",
    "    jkey = f\"{jkey}_{basis}\" if basis is not None else jkey\n",
    "    J_dict = adata.uns[jkey]\n",
    "    c_arr = []\n",
    "    ti_arr = []\n",
    "    for i, reg in enumerate(J_dict[\"regulators\"]):\n",
    "        if regulators is None or reg in regulators:\n",
    "            for j, eff in enumerate(J_dict[\"effectors\"]):\n",
    "                if effectors is None or eff in effectors:\n",
    "                    if i!=j:\n",
    "                        c_arr.append(J_dict[\"jacobian_gene\"][j, i, :])\n",
    "                        ti_arr.append(f\"{eff} wrt. {reg}\")\n",
    "    dyn.pl.multiplot(\n",
    "        lambda c, ti: [dyn.pl.zscatter(adata, color=c, **kwargs), plt.title(ti)],\n",
    "        {\"c\": c_arr, \"ti\": ti_arr},\n",
    "        n_col=2,\n",
    "        subplot_size=(8, 4),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "635698b4",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_jacobian_gene_gene(adata, eff, reg, basis=\"pca\", jkey=\"jacobian\", both=True, regulators=None, effectors=None, **kwargs):\n",
    "    jkey = f\"{jkey}_{basis}\" if basis is not None else jkey\n",
    "    #print(jkey)\n",
    "    J_dict = adata.uns[jkey]\n",
    "    i = np.where(np.array(adata.uns[jkey][\"regulators\"]) == reg)[0][0]\n",
    "    j = np.where(np.array(adata.uns[jkey][\"effectors\"]) == eff)[0][0]\n",
    "    #c_arr = [J_dict[\"jacobian_gene\"][j, i, :], J_dict[\"jacobian_gene\"][i, j, :]]\n",
    "    c_arr = []\n",
    "    ti_arr = []\n",
    "    if both == True:\n",
    "        c_arr = [J_dict[\"jacobian_gene\"][j, i, :], J_dict[\"jacobian_gene\"][i, j, :]]\n",
    "        ti_arr = [f\"{eff} wrt. {reg}\", f\"{reg} wrt. {eff}\"]\n",
    "    else:\n",
    "        c_arr = [J_dict[\"jacobian_gene\"][j, i, :]]\n",
    "        ti_arr = [f\"{eff} wrt. {reg}\"]\n",
    "    dyn.pl.multiplot(\n",
    "        lambda c, ti: [dyn.pl.zscatter(adata, color=c, **kwargs), plt.title(ti)],\n",
    "        {\"c\": c_arr, \"ti\": ti_arr},\n",
    "        n_col=2,\n",
    "        subplot_size=(8, 4),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "38f66f76",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def getNNeighbors(coord, coords, k=150):\n",
    "    nbrs = NearestNeighbors(n_neighbors=k, algorithm='ball_tree').fit(coords)\n",
    "    _, neighs = nbrs.kneighbors(np.atleast_2d(coord))\n",
    "    return neighs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2f195010",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Plots the topography comparing two doses \n",
    "def compare_topography(adata_1, adata_2, geneList, doses):\n",
    "    for gene in geneList:\n",
    "        index = np.where(adata_1.var['var_names'] == gene)[0][0]\n",
    "        _min = np.min(np.concatenate((adata_1.X[:,index].toarray(),adata_2.X[:,index].toarray()), axis=0))\n",
    "        _max = np.max(np.concatenate((adata_1.X[:,index].toarray(),adata_2.X[:,index].toarray()), axis=0))\n",
    "\n",
    "        name = gene + \" gene dose \" + str(doses[0])\n",
    "        adata_1.obs[name] = adata_1.X[:,index].toarray()\n",
    "        dyn.pl.topography(adata_1, basis='umap', color = name, background='white', streamline_color='black', show_legend='on data', frontier=True, vmin=0, vmax=_max)\n",
    "\n",
    "        index = np.where(adata_2.var['var_names'] == gene)[0][0]\n",
    "        name = gene + \" gene dose \" + str(doses[1])\n",
    "        adata_2.obs[name] = adata_2.X[:,index].toarray()\n",
    "        dyn.pl.topography(adata_2, basis='umap', color = name, background='white', streamline_color='black', show_legend='on data', frontier=True, vmin=0, vmax=_max)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "fa9ea8a7",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Plot umap with a specific genes expression values\n",
    "def umap_gene(adata, gene_name):\n",
    "    gene_index = np.where(adata.var['var_names'] == gene_name)[0][0]\n",
    "    gene_exp = adata.X[:,gene_index].toarray().flatten().tolist()\n",
    "    adata.obs[gene_name+'_gene_exp'] = gene_exp\n",
    "    #dyn.pl.umap(adata, color=gene_name+'gene_exp')\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5adda22a",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Plot pca with a specific genes expression values\n",
    "def pca_gene(adata, gene_name):\n",
    "    gene_index = np.where(adata.var['var_names'] == gene_name)[0][0]\n",
    "    gene_exp = adata.X[:,gene_index].toarray().flatten().tolist()\n",
    "    adata.obs[gene_name+'gene_exp'] = gene_exp\n",
    "    #dyn.pl.pca(adata, color=gene_name+'gene_exp')\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ae9d74e5",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Processing data with same UMAP reference\n",
    "# Using a specified gene list\n",
    "def processAdata_genes(adata, processed_adata, genes, dosage, density):\n",
    "    \n",
    "    adata_dose = adata[(dosage == adata.obs['dosage'])]\n",
    "    adata_dose.var.index = adata_dose.var['var_names']\n",
    "    dyn.pp.recipe_monocle(adata_dose, genes_to_use = genes, keep_filtered_genes = False)\n",
    "    \n",
    "    X_umap = processed_adata.obsm['X_umap']\n",
    "    X_pca = processed_adata.obsm['X_pca']\n",
    "    \n",
    "    index = dosage == processed_adata.obs['dosage']\n",
    "    adata_dose.obsm['X_pca'] = X_pca[index]\n",
    "    \n",
    "    dyn.tl.dynamics(adata_dose, model='deterministic', cores=3)\n",
    "    \n",
    "    adata_dose.obsm['X_umap'] = X_umap[index]\n",
    "    \n",
    "    dyn.tl.reduceDimension(adata_dose)\n",
    "    \n",
    "    dyn.tl.cell_velocities(adata_dose, method='pearson', enforce=True, other_kernels_dict={'transform': 'sqrt'}, correct_density = density)\n",
    "    dyn.tl.cell_velocities(adata_dose, basis = 'pca', enforce = True, method='pearson', other_kernels_dict={'transform': 'sqrt'}, correct_density=density)\n",
    "    dyn.vf.VectorField(adata_dose, basis='pca', M=1000)\n",
    "    dyn.vf.VectorField(adata_dose, basis='umap', M=1000)\n",
    "    dyn.vf.topography(adata_dose, basis='pca')\n",
    "    dyn.vf.topography(adata_dose, basis='umap')\n",
    "    \n",
    "    return adata_dose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fd8310c1",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Processing data with individiual UMAP reference\n",
    "# Using a specified gene list\n",
    "def processAdata_genes2(adata, processed_adata, genes, dosage, density):\n",
    "    \n",
    "    adata_dose = adata[(dosage == adata.obs['dosage'])]\n",
    "    adata_dose.var.index = adata_dose.var['var_names']\n",
    "    dyn.pp.recipe_monocle(adata_dose, genes_to_use = genes, keep_filtered_genes = False)\n",
    "    \n",
    "    #X_umap = processed_adata.obsm['X_umap']\n",
    "    #X_pca = processed_adata.obsm['X_pca']\n",
    "    \n",
    "    #index = dosage == processed_adata.obs['dosage']\n",
    "    #adata_dose.obsm['X_pca'] = X_pca[index]\n",
    "    \n",
    "    dyn.tl.dynamics(adata_dose, model='deterministic', cores=3)\n",
    "    \n",
    "    #adata_dose.obsm['X_umap'] = X_umap[index]\n",
    "    \n",
    "    dyn.tl.reduceDimension(adata_dose)\n",
    "    \n",
    "    dyn.tl.cell_velocities(adata_dose, method='pearson', enforce=True, other_kernels_dict={'transform': 'sqrt'}, correct_density = density)\n",
    "    dyn.tl.cell_velocities(adata_dose, basis = 'pca', enforce = True, method='pearson', other_kernels_dict={'transform': 'sqrt'}, correct_density=density)\n",
    "    dyn.vf.VectorField(adata_dose, basis='pca', M=1000)\n",
    "    dyn.vf.VectorField(adata_dose, basis='umap', M=1000)\n",
    "    dyn.vf.topography(adata_dose, basis='pca')\n",
    "    dyn.vf.topography(adata_dose, basis='umap')\n",
    "    \n",
    "    return adata_dose"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f89a437",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Saving and reloading adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e4461d4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# B is assumed to be a numpy array\n",
    "def saveas_R_file(B, filename):\n",
    "    rpy2.robjects.numpy2ri.activate()\n",
    "    nr,nc = B.shape\n",
    "    Br = ro.r.matrix(B, nrow=nr, ncol=nc)\n",
    "    ro.r.assign(\"B\", Br)\n",
    "    ro.r(\"save(B, file='\"+filename+\".Rdata')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "e73d231a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_vector_field(adata_list, processed_adata, file_name, dict_name):\n",
    "    for i in range(len(adata_list)):\n",
    "        dose = np.unique(adata_list[i].obs['dosage'])[0]\n",
    "        print(dose)\n",
    "        temp_name = file_name + str(dose) + \".h5ad\"\n",
    "        #temp = processed_adata[(dose == processed_adata.obs['dosage'])]\n",
    "        #adata_list[i].var.index = temp.var.index\n",
    "\n",
    "        adata_list[i].var.index = np.arange(adata_list[i].shape[1])\n",
    "        adata_list[i].var = adata_list[i].var.fillna(0)\n",
    "        file = open(dict_name + str(dose) + '.pkl', 'wb')\n",
    "        pickle.dump(adata_list[i].uns, file)\n",
    "        file.close()\n",
    "\n",
    "        temp_adata = adata_list[i].copy()\n",
    "        temp_adata.uns = {}\n",
    "        temp_adata.write_h5ad(filename=temp_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee088027",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_processed_adata(adata, processed_adata, file_name, dict_name):\n",
    "    #dose = np.unique(adata_list[i].obs['dosage'])[0]\n",
    "    #print(dose)\n",
    "    temp_name = file_name + \".h5ad\"\n",
    "    #temp = processed_adata[(dose == processed_adata.obs['dosage'])]\n",
    "    #processed_adata.var.index = adata.var.index\n",
    "\n",
    "    file = open(dict_name + '.pkl', 'wb')\n",
    "    temp_dict = {}\n",
    "    for key in processed_adata.uns.keys():\n",
    "        print(key)\n",
    "        if key != 'neighbors':\n",
    "            temp_dict[key] = processed_adata.uns[key]\n",
    "    print(\"done loading\")\n",
    "    #pickle.dump(processed_adata.uns, file)\n",
    "    pickle.dump(temp_dict, file)\n",
    "    file.close()\n",
    "\n",
    "    temp_adata = processed_adata.copy()\n",
    "    temp_adata.uns = {}\n",
    "    #adata_list[i].uns = {}\n",
    "    temp_adata.write_h5ad(filename=temp_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b956f0e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_processed_adata2(adata, processed_adata, file_name, dict_name, umap_model_name):\n",
    "    #dose = np.unique(adata_list[i].obs['dosage'])[0]\n",
    "    #print(dose)\n",
    "    temp_name = file_name + \".h5ad\"\n",
    "    #temp = processed_adata[(dose == processed_adata.obs['dosage'])]\n",
    "    #processed_adata.var.index = adata.var.index\n",
    "\n",
    "    #processed_adata.uns['umap_fit']['fit'].save(umap_model_name)\n",
    "\n",
    "    file = open(dict_name + '.pkl', 'wb')\n",
    "    temp_dict = {}\n",
    "    for key in processed_adata.uns.keys():\n",
    "        #if key!='neighbors' or key!='umap_fit':\n",
    "        #   temp_dict[key] = processed_adata.uns[key]\n",
    "        if 'neighbors' not in key:\n",
    "            temp_dict[key] = processed_adata.uns[key]\n",
    "        \n",
    "    #pickle.dump(processed_adata.uns, file)\n",
    "    pickle.dump(temp_dict, file)\n",
    "    file.close()\n",
    "\n",
    "    temp_adata = processed_adata.copy()\n",
    "    temp_adata.uns = {}\n",
    "    #adata_list[i].uns = {}\n",
    "    temp_adata.write_h5ad(filename=temp_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e4ce1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# INPUTS:\n",
    "#   length = indicates how many doses need to loaded (defualt = 8 which means all doses are going to be loaded)\n",
    "def reload_data(file_name, dict_name, length=8):\n",
    "    reloaded = []\n",
    "    doses = [0, 12, 25, 50, 100, 200, 400, 800]\n",
    "    for i in range(length):\n",
    "        temp_name = file_name + str(doses[i]) + \".h5ad\"\n",
    "        temp = dyn.read_h5ad(temp_name)\n",
    "        loaded_dict = pd.read_pickle(dict_name + str(doses[i]) + '.pkl')\n",
    "        temp.uns = loaded_dict\n",
    "        reloaded.append(temp.copy())\n",
    "    return reloaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f2d8849e-adac-47a5-9ff1-7314d5329c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# INPUTS:\n",
    "#   length = indicates how many doses need to loaded (defualt = 8 which means all doses are going to be loaded)\n",
    "def reload_data2(file_name, dict_name, length=8):\n",
    "    reloaded = []\n",
    "    doses = [0, 12, 25, 50, 100, 200, 400, 800]\n",
    "    for i in range(length):\n",
    "        temp_name = file_name + str(doses[i]) + \".h5ad\"\n",
    "        temp = dyn.read_h5ad(temp_name)\n",
    "        #loaded_dict = pd.read_pickle(dict_name + str(doses[i]) + '.pkl')\n",
    "        #temp.uns = loaded_dict\n",
    "        reloaded.append(temp.copy())\n",
    "    return reloaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5e0c6eb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reload_processed_data(file_name, dict_name):\n",
    "    \n",
    "    temp_name = file_name + \".h5ad\"\n",
    "    temp = dyn.read_h5ad(temp_name)\n",
    "    loaded_dict = pd.read_pickle(dict_name + '.pkl')\n",
    "    temp.uns = loaded_dict\n",
    "    #temp.uns['umap_fit'] = load_ParametricUMAP(umap_model_name)\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c998b8ae-d2f4-49a0-a7bd-6dacaf69e813",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reload_processed_data_new(file_name, dict_name):\n",
    "    \n",
    "    temp_name = file_name + \".h5ad\"\n",
    "    temp = dyn.read_h5ad(temp_name)\n",
    "    #loaded_dict = pd.read_pickle(dict_name + '.pkl')\n",
    "    #temp.uns = loaded_dict\n",
    "    #temp.uns['umap_fit'] = load_ParametricUMAP(umap_model_name)\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b68d0fd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reload_processed_data2(file_name, dict_name, umap_model_name):\n",
    "    \n",
    "    temp_name = file_name + \".h5ad\"\n",
    "    temp = dyn.read_h5ad(temp_name)\n",
    "    loaded_dict = pd.read_pickle(dict_name + '.pkl')\n",
    "    temp.uns = loaded_dict\n",
    "    temp.uns['umap_fit'] = load_ParametricUMAP(umap_model_name)\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0d59592",
   "metadata": {
    "heading_collapsed": true,
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Fixed points"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b86be8ec",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def getFixedPoints(adata, basis):\n",
    "    basis_key = \"VecFld_\" + basis\n",
    "    if basis == 'umap':\n",
    "        if adata.uns[basis_key]['Xss'].shape[0] > 0:\n",
    "            all_fps = np.hstack((adata.uns[basis_key]['Xss'][:,0:2], adata.uns[basis_key]['ftype'][:,None], adata.uns[basis_key]['confidence'][:,None]))\n",
    "        else:\n",
    "            return [],[],[],[]\n",
    "    else:\n",
    "        if adata.uns[basis_key]['X'].shape[1] == 2:\n",
    "            if adata.uns[basis_key]['Xss'].shape[0] > 0:\n",
    "                all_fps = np.hstack((adata.uns[basis_key]['Xss'][:,0:2], adata.uns[basis_key]['ftype'][:,None], adata.uns[basis_key]['confidence'][:,None]))\n",
    "            else:\n",
    "                return [],[],[],[]\n",
    "        else:\n",
    "            if adata.uns[basis_key]['Xss'].shape[0] > 0:\n",
    "                all_fps = np.hstack((adata.uns[basis_key]['Xss'][:,0:2], adata.uns[basis_key]['ftype'][:,None]))\n",
    "            else:\n",
    "                return [],[],[],[]\n",
    "    stable = all_fps[np.where(all_fps[:,2] == -1)]\n",
    "    unstable = all_fps[np.where(all_fps[:,2] == 1)]\n",
    "    saddle = all_fps[np.where(all_fps[:,2] == 0)]\n",
    "    return all_fps, stable, unstable, saddle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "08fd0d3f",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Plots the fixed points for a specific dose\n",
    "# adata - the current dose \n",
    "# adata_forward - the dose after the current dose\n",
    "# adata_backward - the dose before the current dose\n",
    "# stable_points, unstable_points, saddle_points = fixed points of the adata current dose\n",
    "def plot_fp(adata, adata_forward, adata_backward, stable_points, unstable_points, saddle_points, fileName):\n",
    "    dose = adata.obs['dosage'][0]\n",
    "    if adata_forward is not None:\n",
    "        plt.scatter(adata_forward.obsm['X_umap'][:,0], adata_forward.obsm['X_umap'][:,1], c='gainsboro', label=\"next dose\")\n",
    "    if adata_backward is not None:\n",
    "        plt.scatter(adata_backward.obsm['X_umap'][:,0], adata_backward.obsm['X_umap'][:,1], c='silver', label=\"previous dose\")\n",
    "    plt.scatter(adata.obsm['X_umap'][:,0], adata.obsm['X_umap'][:,1], c='mediumpurple', label=\"current dose\")\n",
    "    #plt.scatter(stable_points[:,0], stable_points[:,1], c='r', label='stable')\n",
    "    plt.scatter(unstable_points[:,0], unstable_points[:,1], c='violet', label='unstable')\n",
    "    plt.scatter(saddle_points[:,0], saddle_points[:,1], c='blue', label='saddle')\n",
    "    plt.scatter(stable_points[:,0], stable_points[:,1], c='r', label='stable')\n",
    "    plt.legend(bbox_to_anchor=(1.35,1), loc=\"upper right\")\n",
    "    plt.xlabel(\"UMAP 0\")\n",
    "    plt.ylabel(\"UMAP 1\")\n",
    "    plt.title(\"TGFB dose \" + dose + \" fixed points\")\n",
    "    #plt.savefig(fileName, bbox_inches='tight', dpi=300)\n",
    "    #plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b86955d5",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def gene_acceleration(adata, gene_name):\n",
    "    gene_index = np.where(adata.var['var_names'] == gene_name)[0][0]\n",
    "    gene_acc = adata.layers['acceleration'][:,gene_index].tolist()\n",
    "    adata.obs[gene_name+'_accel'] = gene_acc\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e0ca0642",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def gene_curvature(adata, gene_name):\n",
    "    gene_index = np.where(adata.var['var_names'] == gene_name)[0][0]\n",
    "    gene_cur = adata.layers['curvature'][:,gene_index].tolist()\n",
    "    adata.obs[gene_name+'_curv'] = gene_cur\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4b4f0da2",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def gene_velocity(adata, gene_name):\n",
    "    gene_index = np.where(adata.var['var_names'] == gene_name)[0][0]\n",
    "    gene_vel = adata.layers['velocity_S'][:,gene_index].toarray().flatten().tolist()\n",
    "    adata.obs[gene_name+'_vel'] = gene_vel\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "70530dd1",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def gene_exp(adata, gene_name):\n",
    "    gene_index = np.where(adata.var['var_names'] == gene_name)[0][0]\n",
    "    gene_exp = adata.X[:,gene_index].toarray().flatten().tolist()\n",
    "    adata.obs[gene_name+'_gene_exp'] = gene_exp\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5dbc1d85",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def connect_ind_same_UMAP(adata_ind, adata_same, ind_stable_points, same_stable_points, fileName):\n",
    "    dose = np.unique(adata_ind.obs['dosage'])[0]\n",
    "    print(\"Dose \", dose, \":\")\n",
    "    \n",
    "    new_coords = []\n",
    "    \n",
    "    plt.figure(figsize=(15,3))\n",
    "    plt.subplot(1, 3, 1)\n",
    "    plt.scatter(adata_same.obsm['X_umap'][:,0], adata_same.obsm['X_umap'][:,1], c=\"lightgrey\")\n",
    "    dyn.pl.zstreamline(adata_same)\n",
    "    plt.scatter(same_stable_points[:,0], same_stable_points[:,1], c='r', label='stable')\n",
    "    plt.title(\"Same UMAP ref - original fp\")\n",
    "    \n",
    "    plt.subplot(1, 3, 2)\n",
    "    plt.scatter(adata_same.obsm['X_umap'][:,0], adata_same.obsm['X_umap'][:,1], c=\"lightgrey\")\n",
    "    dyn.pl.zstreamline(adata_same)\n",
    "    \n",
    "    # Plotting the individual fixed points onto the same UMAP reference\n",
    "    for i in range(len(ind_stable_points)):\n",
    "        index = getNNeighbors([ind_stable_points[i][0], ind_stable_points[i][1]], adata_ind.obsm['X_umap'], 1)[0][0]\n",
    "        fp_coords = ind_stable_points[i]\n",
    "        ind_coords = adata_ind.obsm['X_umap'][index]\n",
    "        dist = np.sqrt((fp_coords[0] - ind_coords[0])**2 + (fp_coords[1] - ind_coords[1])**2)\n",
    "        if dist < 1:\n",
    "            new_coords.append(ind_coords.tolist())\n",
    "            plt.scatter(adata_same.obsm['X_umap'][index][0], adata_same.obsm['X_umap'][index][1], label=\"stable point\", s=60)\n",
    "    #plt.legend(loc=\"upper right\", bbox_to_anchor=(1.3, 1))\n",
    "    plt.title(\"Ind fixed points mapped onto same UMAP ref\")\n",
    "    \n",
    "    # Plotting the same UMAP reference onto the individual fixed points\n",
    "    plt.subplot(1, 3, 3)\n",
    "    plt.scatter(adata_ind.obsm['X_umap'][:,0], adata_ind.obsm['X_umap'][:,1], c=\"lightgrey\")\n",
    "    dyn.pl.zstreamline(adata_ind)\n",
    "    for i in range(len(new_coords)):\n",
    "        plt.scatter(new_coords[i][0], new_coords[i][1], label=\"stable point\", s=60)\n",
    "    #plt.legend(loc=\"upper right\", bbox_to_anchor=(1.3, 1))\n",
    "    plt.title(\"Ind fixed points mapped onto ind UMAP ref\")\n",
    "    plt.savefig(fileName+\"mapping_\"+str(dose), dpi=300)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ebfa1996",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def connect_ind_same_UMAP_overlap(adata, adata_ind, adata_same, ind_stable_points, same_stable_points, fileName):\n",
    "    dose = np.unique(adata_ind.obs['dosage'])[0]\n",
    "    print(\"Dose \", dose, \":\")\n",
    "    \n",
    "    new_coords = [] #filtered fixed points for the individual UMAP ref \n",
    "    \n",
    "    # Plots the original fixed points on the same UMAP ref\n",
    "    plt.figure(figsize=(15,4))\n",
    "    plt.subplot(1, 3, 1)\n",
    "    plt.scatter(adata.obsm['X_umap'][:,0], adata.obsm['X_umap'][:,1], c = 'lavender')\n",
    "    plt.scatter(adata_same.obsm['X_umap'][:,0], adata_same.obsm['X_umap'][:,1], c=\"silver\", alpha=0.4)\n",
    "    dyn.pl.zstreamline(adata_same)\n",
    "    plt.scatter(same_stable_points[:,0], same_stable_points[:,1], c='r', label='stable')\n",
    "    #plt.title(\"Same UMAP ref - original fp\")\n",
    "    \n",
    "    #Plots the overlapping fixed points from the ind and same ref as bigger dots \n",
    "    #Plots non overlapping fixed points as smaller dots\n",
    "    plt.subplot(1, 3, 2)\n",
    "    plt.scatter(adata.obsm['X_umap'][:,0], adata.obsm['X_umap'][:,1], c = 'lavender')\n",
    "    plt.scatter(adata_same.obsm['X_umap'][:,0], adata_same.obsm['X_umap'][:,1], c=\"silver\", alpha=0.4)\n",
    "    dyn.pl.zstreamline(adata_same)\n",
    "    \n",
    "    mapped_points = [] #tracks the mapped fixed points ind --> same UMAP ref\n",
    "    # Plots the individual fixed points onto the same UMAP reference\n",
    "    for i in range(len(ind_stable_points)):\n",
    "        index = getNNeighbors([ind_stable_points[i][0], ind_stable_points[i][1]], adata_ind.obsm['X_umap'], 1)[0][0]\n",
    "        fp_coords = ind_stable_points[i]\n",
    "        ind_coords = adata_ind.obsm['X_umap'][index]\n",
    "        dist = np.sqrt((fp_coords[0] - ind_coords[0])**2 + (fp_coords[1] - ind_coords[1])**2)\n",
    "        if dist < 1:\n",
    "            new_coords.append(ind_coords.tolist())\n",
    "            #print([adata_same.obsm['X_umap'][index][0], adata_same.obsm['X_umap'][index][1]])\n",
    "            mapped_points.append([adata_same.obsm['X_umap'][index][0], adata_same.obsm['X_umap'][index][1]])\n",
    "            plt.scatter(adata_same.obsm['X_umap'][index][0], adata_same.obsm['X_umap'][index][1], label=\"stable point\", s=60)\n",
    "    \n",
    "    to_keep = [] #fixed points from the same UMAP ref that are inside the boundary \n",
    "    #Filters out fixed points from the same UMAP ref that are outside the boundary\n",
    "    for i in range(len(same_stable_points)):\n",
    "        index = getNNeighbors([same_stable_points[i][0], same_stable_points[i][1]], adata_same.obsm['X_umap'], 1)[0][0]\n",
    "        fp_coords = same_stable_points[i]\n",
    "        ind_coords = adata_same.obsm['X_umap'][index]\n",
    "        dist = np.sqrt((fp_coords[0] - ind_coords[0])**2 + (fp_coords[1] - ind_coords[1])**2)\n",
    "        if dist < 1:\n",
    "            to_keep.append(fp_coords[0:2].tolist())\n",
    "            plt.scatter(adata_same.obsm['X_umap'][index][0], adata_same.obsm['X_umap'][index][1], c='r', s=60)\n",
    "    #plt.title(\"Ind + same fp mapped onto same UMAP ref\")\n",
    "\n",
    "    plt.subplot(1, 3, 3)\n",
    "    plt.scatter(adata.obsm['X_umap'][:,0], adata.obsm['X_umap'][:,1], c = 'lavender')\n",
    "    plt.scatter(adata_same.obsm['X_umap'][:,0], adata_same.obsm['X_umap'][:,1], c=\"silver\", alpha=0.4)\n",
    "    dyn.pl.zstreamline(adata_same)\n",
    "    \n",
    "    for i in range(len(mapped_points)):\n",
    "        plt.scatter(mapped_points[i][0], mapped_points[i][1], label=\"stable point\", s=60, c='b')\n",
    "    \n",
    "    # checks if ind + same UMAP fixed points overlap - plots as a bigger point\n",
    "    # plots the non overlapping fixed points as smaller points\n",
    "    for i in range(len(to_keep)):\n",
    "        index = getNNeighbors(to_keep[i], mapped_points, 1)[0][0]\n",
    "        cur_coord = to_keep[i]\n",
    "        other_coord = mapped_points[index]\n",
    "        dist = np.sqrt((cur_coord[0] - other_coord[0])**2 + (cur_coord[1] - other_coord[1])**2)\n",
    "        if dist < 2:\n",
    "            plt.scatter(cur_coord[0], cur_coord[1], label=\"stable point\", s=300, c='greenyellow')\n",
    "        else:\n",
    "            plt.scatter(cur_coord[0], cur_coord[1], label=\"stable point\", s=60, c='r')\n",
    "    #print(mapped_points)  \n",
    "    #for i in range(len(mapped_points)):\n",
    "    #    plt.scatter(mapped_points[i][0], mapped_points[i][1], label=\"stable point\", s=60, c='b')\n",
    "        \n",
    "    #plt.title(\"Ind + same fp mapped onto same UMAP ref\")\n",
    "    \n",
    "    \n",
    "    # Plotting the same UMAP reference onto the individual fixed points\n",
    "    #plt.subplot(2, 2, 4)\n",
    "    #plt.scatter(adata_ind.obsm['X_umap'][:,0], adata_ind.obsm['X_umap'][:,1], c=\"silver\", alpha=0.4)\n",
    "    #dyn.pl.zstreamline(adata_ind)\n",
    "    #for i in range(len(new_coords)):\n",
    "    #    plt.scatter(new_coords[i][0], new_coords[i][1], label=\"stable point\", s=60)\n",
    "    #plt.title(\"Ind fixed points mapped onto ind UMAP ref\")\n",
    "    #plt.savefig(fileName+\"mapping_overlapping_\"+str(dose), dpi=300)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d8156960",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def connect_ind_same_UMAP_overlap2(adata, adata_ind, adata_same, ind_stable_points, same_stable_points, fileName):\n",
    "    dose = np.unique(adata_ind.obs['dosage'])[0]\n",
    "    print(\"Dose \", dose, \":\")\n",
    "    \n",
    "    new_coords = [] #filtered fixed points for the individual UMAP ref \n",
    "    \n",
    "    # Plots the original fixed points on the same UMAP ref\n",
    "#     plt.figure(figsize=(15,4))\n",
    "#     plt.subplot(1, 3, 1)\n",
    "#     plt.scatter(adata.obsm['X_umap'][:,0], adata.obsm['X_umap'][:,1], c = 'lavender')\n",
    "#     plt.scatter(adata_same.obsm['X_umap'][:,0], adata_same.obsm['X_umap'][:,1], c=\"silver\", alpha=0.4)\n",
    "#     dyn.pl.zstreamline(adata_same)\n",
    "#     plt.scatter(same_stable_points[:,0], same_stable_points[:,1], c='r', label='stable')\n",
    "#     #plt.title(\"Same UMAP ref - original fp\")\n",
    "    \n",
    "    #Plots the overlapping fixed points from the ind and same ref as bigger dots \n",
    "    #Plots non overlapping fixed points as smaller dots\n",
    "    #plt.subplot(1, 3, 2)\n",
    "#     plt.scatter(adata.obsm['X_umap'][:,0], adata.obsm['X_umap'][:,1], c = 'lavender')\n",
    "#     plt.scatter(adata_same.obsm['X_umap'][:,0], adata_same.obsm['X_umap'][:,1], c=\"silver\", alpha=0.4)\n",
    "#     dyn.pl.zstreamline(adata_same)\n",
    "    \n",
    "    mapped_points = [] #tracks the mapped fixed points ind --> same UMAP ref\n",
    "    # Plots the individual fixed points onto the same UMAP reference\n",
    "    for i in range(len(ind_stable_points)):\n",
    "        index = getNNeighbors([ind_stable_points[i][0], ind_stable_points[i][1]], adata_ind.obsm['X_umap'], 1)[0][0]\n",
    "        fp_coords = ind_stable_points[i]\n",
    "        ind_coords = adata_ind.obsm['X_umap'][index]\n",
    "        dist = np.sqrt((fp_coords[0] - ind_coords[0])**2 + (fp_coords[1] - ind_coords[1])**2)\n",
    "        if dist < 1:\n",
    "            new_coords.append(ind_coords.tolist())\n",
    "            #print([adata_same.obsm['X_umap'][index][0], adata_same.obsm['X_umap'][index][1]])\n",
    "            mapped_points.append([adata_same.obsm['X_umap'][index][0], adata_same.obsm['X_umap'][index][1]])\n",
    "            #plt.scatter(adata_same.obsm['X_umap'][index][0], adata_same.obsm['X_umap'][index][1], label=\"stable point\", s=60)\n",
    "    \n",
    "    to_keep = [] #fixed points from the same UMAP ref that are inside the boundary \n",
    "    #Filters out fixed points from the same UMAP ref that are outside the boundary\n",
    "    for i in range(len(same_stable_points)):\n",
    "        index = getNNeighbors([same_stable_points[i][0], same_stable_points[i][1]], adata_same.obsm['X_umap'], 1)[0][0]\n",
    "        fp_coords = same_stable_points[i]\n",
    "        ind_coords = adata_same.obsm['X_umap'][index]\n",
    "        dist = np.sqrt((fp_coords[0] - ind_coords[0])**2 + (fp_coords[1] - ind_coords[1])**2)\n",
    "        if dist < 1:\n",
    "            to_keep.append(fp_coords[0:2].tolist())\n",
    "            #plt.scatter(adata_same.obsm['X_umap'][index][0], adata_same.obsm['X_umap'][index][1], c='r', s=60)\n",
    "    #plt.title(\"Ind + same fp mapped onto same UMAP ref\")\n",
    "\n",
    "    #plt.subplot(1, 3, 3)\n",
    "    plt.scatter(adata.obsm['X_umap'][:,0], adata.obsm['X_umap'][:,1], c = 'lavender')\n",
    "    plt.scatter(adata_same.obsm['X_umap'][:,0], adata_same.obsm['X_umap'][:,1], c=\"silver\", alpha=0.4)\n",
    "    dyn.pl.zstreamline(adata_same)\n",
    "    \n",
    "    for i in range(len(mapped_points)):\n",
    "        plt.scatter(mapped_points[i][0], mapped_points[i][1], label=\"stable point\", s=60, c='b')\n",
    "    \n",
    "    # checks if ind + same UMAP fixed points overlap - plots as a bigger point\n",
    "    # plots the non overlapping fixed points as smaller points\n",
    "    for i in range(len(to_keep)):\n",
    "        index = getNNeighbors(to_keep[i], mapped_points, 1)[0][0]\n",
    "        cur_coord = to_keep[i]\n",
    "        other_coord = mapped_points[index]\n",
    "        dist = np.sqrt((cur_coord[0] - other_coord[0])**2 + (cur_coord[1] - other_coord[1])**2)\n",
    "        if dist < 2:\n",
    "            plt.scatter(cur_coord[0], cur_coord[1], label=\"stable point\", s=300, c='greenyellow')\n",
    "        else:\n",
    "            plt.scatter(cur_coord[0], cur_coord[1], label=\"stable point\", s=60, c='r')\n",
    "    #print(mapped_points)  \n",
    "    #for i in range(len(mapped_points)):\n",
    "    #    plt.scatter(mapped_points[i][0], mapped_points[i][1], label=\"stable point\", s=60, c='b')\n",
    "        \n",
    "    #plt.title(\"Ind + same fp mapped onto same UMAP ref\")\n",
    "    \n",
    "    \n",
    "    # Plotting the same UMAP reference onto the individual fixed points\n",
    "    #plt.subplot(2, 2, 4)\n",
    "    #plt.scatter(adata_ind.obsm['X_umap'][:,0], adata_ind.obsm['X_umap'][:,1], c=\"silver\", alpha=0.4)\n",
    "    #dyn.pl.zstreamline(adata_ind)\n",
    "    #for i in range(len(new_coords)):\n",
    "    #    plt.scatter(new_coords[i][0], new_coords[i][1], label=\"stable point\", s=60)\n",
    "    #plt.title(\"Ind fixed points mapped onto ind UMAP ref\")\n",
    "    plt.savefig(fileName+\"mapping_overlapping_\"+str(dose), dpi=300)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "963b6bcf",
   "metadata": {
    "heading_collapsed": true,
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Plotting trajectory and gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d5b89b64",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Using the start and end coords the closest cell is obtained\n",
    "# LAP is performed and the resulting trajectory is plotted on the PCA reference space\n",
    "\n",
    "def plot_single_trajectory(start_coord, end_coord, adata, num_points):\n",
    "    start_index = getNNeighbors(start_coord, adata.obsm['X_pca'][:,0:2], k=1)\n",
    "    end_index = getNNeighbors(end_coord, adata.obsm['X_pca'][:,0:2], k=1)\n",
    "\n",
    "    start_cell = adata.obsm['X_pca'][start_index[0][0]]\n",
    "    end_cell = adata.obsm['X_pca'][end_index[0][0]]\n",
    "\n",
    "    adata.obsm['X_pca'] = adata.obsm['X_pca'][:,0:2]\n",
    "    adata.uns['PCs'] = adata.uns['PCs'][:,0:2]\n",
    "\n",
    "    dyn.pd.least_action(adata, start_index[0], end_index[0], n_points=num_points)\n",
    "    \n",
    "    path_sol = adata.uns['LAP_pca']['prediction'][0]\n",
    "    \n",
    "    plt.scatter(adata.obsm[\"X_pca\"][:,0], adata.obsm[\"X_pca\"][:,1], color = \"lightgrey\")\n",
    "    plt.scatter(path_sol[:,0], path_sol[:,1], c=path_sol[:,0])\n",
    "    plt.scatter(start_cell[0], start_cell[1], marker = \"*\", color = \"red\", label=\"start\")\n",
    "    plt.scatter(end_cell[0], end_cell[1], marker = \"*\", color = \"yellow\", label = \"end\")\n",
    "    plt.legend(loc=\"upper left\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8af4fc86",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def pca_to_expr(X, PCs, mean=0, func=None):\n",
    "    # reverse project from PCA back to raw expression space\n",
    "    if PCs.shape[1] == X.shape[1]:\n",
    "        exprs = X @ PCs.T + mean\n",
    "        if func is not None:\n",
    "            exprs = func(exprs)\n",
    "    else:\n",
    "        raise Exception(\"PCs dim 1 (%d) does not match X dim 1 (%d).\" % (PCs.shape[1], X.shape[1]))\n",
    "    return exprs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c11ff691",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def group_score(adata, expr_output, all_genes, gene_list: list):\n",
    "\n",
    "    tmp = all_genes.intersection(gene_list)\n",
    "    #print(tmp.shape)\n",
    "    # use indices\n",
    "    intersect_genes = [all_genes.get_loc(i) for i in tmp]\n",
    "    #print(intersect_genes)\n",
    "\n",
    "    if len(intersect_genes) == 0:\n",
    "        raise Exception(f\"your adata doesn't have any gene from the gene_list {gene_list}.\")\n",
    "\n",
    "    #print(intersect_genes)\n",
    "    expression_matrix = expr_output[:, intersect_genes]\n",
    "    expression_matrix = log1p_(adata, expression_matrix)\n",
    "    \n",
    "    if issparse(expression_matrix):\n",
    "        expression_matrix.data = np.log1p(expression_matrix.data)\n",
    "        scores = expression_matrix.sum(1).A1\n",
    "    else:\n",
    "        scores = np.log1p(expression_matrix).sum(1)\n",
    "\n",
    "    scores = (scores - scores.mean()) / scores.std()\n",
    "\n",
    "    return scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1f7b3ba3",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def batch_group_score(adata, expr_output, all_genes, gene_lists: list):\n",
    "    batch_scores = collections.OrderedDict()\n",
    "    for gene_list in gene_lists:\n",
    "        batch_scores[gene_list] = group_score(adata, expr_output, all_genes, gene_lists[gene_list])\n",
    "    return batch_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0ae6969e",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def log1p_(adata, X_data):\n",
    "    if \"norm_method\" not in adata.uns[\"pp\"].keys():\n",
    "        return X_data\n",
    "    else:\n",
    "        if adata.uns[\"pp\"][\"norm_method\"] is None:\n",
    "            if issparse(X_data):\n",
    "                X_data.data = np.log1p(X_data.data)\n",
    "            else:\n",
    "                X_data = np.log1p(X_data)\n",
    "\n",
    "        return X_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aaf004f9",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def get_LAP(start_coord, end_coord, adata, num_points, dim=30):\n",
    "    #start_index = getNNeighbors(start_coord, adata.obsm['X_pca'][:,dim], k=1)\n",
    "    #end_index = getNNeighbors(end_coord, adata.obsm['X_pca'][:,dim], k=1)\n",
    "\n",
    "    #start_cell = adata.obsm['X_pca'][start_index[0][0]]\n",
    "    #end_cell = adata.obsm['X_pca'][end_index[0][0]]\n",
    "\n",
    "    adata.obsm['X_pca'] = adata.obsm['X_pca'][:,:dim]\n",
    "    adata.uns['PCs'] = adata.uns['PCs'][:,:dim]\n",
    "\n",
    "    #traj = dyn.pd.least_action(adata, start_index[0], end_index[0], n_points=num_points)\n",
    "    traj = dyn.pd.least_action(adata, start_coord, end_coord, n_points=num_points)\n",
    "    \n",
    "    path_sol = adata.uns['LAP_pca']['prediction'][0]\n",
    "    \n",
    "    x_pca = adata.uns['LAP_pca']['prediction'][0]\n",
    "    PCs = adata.uns['PCs']\n",
    "    mean = adata.uns['pca_mean']\n",
    "    #expr_output = pca_to_expr(x_pca, PCs, mean, np.expm1)\n",
    "    expr_output = pca_to_expr(x_pca, PCs, mean, None)\n",
    "    \n",
    "    all_genes = adata.var.index[np.where(adata.var['use_for_dynamics'] == True)[0]]\n",
    "    #display(pd.DataFrame(all_genes))\n",
    "\n",
    "    return path_sol, expr_output, all_genes, traj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ddd9e608",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# def plot_single_trajectory(start_coord, end_coord, adata):\n",
    "#     path_sol = dyn.pd.least_action(adata, start_coord, end_coord)\n",
    "#     plt.scatter(adata.obsm[\"X_umap\"][:,0], adata.obsm[\"X_umap\"][:,1], color = \"lightgrey\")\n",
    "#     plt.scatter(path_sol[:,0], path_sol[:,1], c=path_sol[:,0])\n",
    "#     plt.scatter(start_coord[0], start_coord[1], marker = \"*\", color = \"red\", label=\"start\")\n",
    "#     plt.scatter(end_coord[0], end_coord[1], marker = \"*\", color = \"black\", label = \"end\")\n",
    "#     plt.legend(loc=\"upper left\")\n",
    "#     #plt.savefig('Plots/EMT_trajectory/DE_genes_top_path/top_trajectory', dpi=300)\n",
    "#     plt.show()\n",
    "#     return path_sol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a5abc94",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Using the start and end coords the closest cell is obtained\n",
    "# LAP is performed and the resulting trajectory is plotted on the PCA reference space\n",
    "# Only varies the starting point - not the end point\n",
    "def plot_multiple_trajectories(start_coord, end_coord, adata, number_of_cells):\n",
    "    start_indices = getNNeighbors(start_coord, adata.obsm['X_umap'], k=number_of_cells)[0]\n",
    "    #end_indices = getNNeighbors([8.5,1], adata.obsm['X_umap'], k=10)[0]\n",
    "    \n",
    "    for i in range(len(start_indices)):\n",
    "        start = adata.obsm[\"X_umap\"][start_indices[i], :]\n",
    "        print(\"Starting point: \", start)\n",
    "        plot_single_trajectory(start, end_coord, adata)\n",
    "        #dyn.pd.least_action(adata, start, end_coord)\n",
    "        #path_sol = adata.uns['LAP_pca']['prediction'][0]\n",
    "\n",
    "#         plt.scatter(adata.obsm[\"X_pca\"][:,0], adata.obsm[\"X_\"][:,1], color = \"lightgrey\")\n",
    "#         plt.scatter(path_sol[:,0], path_sol[:,1], c=path_sol[:,0])\n",
    "#         plt.scatter(adata.obsm[\"X_umap\"][start_indices[i],0], adata.obsm[\"X_umap\"][start_indices[i],1], marker = \"*\", color = \"red\", label=\"start\")\n",
    "#         #plt.scatter(adata.obsm[\"X_umap\"][end_coord,0], adata.obsm[\"X_umap\"][end_coord,1], marker = \"*\", color = \"black\", label = \"end\")\n",
    "#         plt.scatter(end_coord[0], end_coord[1], marker = \"*\", color = \"black\", label = \"end\")\n",
    "#         #plt.title(\"TGFB dose 12.5 - least action path\")\n",
    "#         plt.legend(loc=\"upper left\")\n",
    "#         filename = \"Plots/EMT_trajectory/Trajectories/traj_\"+str(i)\n",
    "#         plt.savefig(filename, dpi=300)\n",
    "#         plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "94f665e8",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Parameters: \n",
    "#     num_points: number of points for each trajectory\n",
    "#     num_traj: number of trajectories to generate\n",
    "# Returns:\n",
    "#     pathways: 2D list containing the coordinates for each trajectory\n",
    "#     progression: 2D list containing the cell cycle categories for each trajectory\n",
    "\n",
    "def cell_progression(start_coord, end_coord, adata, num_points, num_traj, genes):\n",
    "    phases = np.array(['G1-S', 'S', 'G2-M', 'M', 'M-G1'], dtype=object)\n",
    "    myColors = ((255/256,91/256,165/256),(219/256,73/256,172/256), (176/256,124/256,218/256), (107/256,167/256,214/256),(100/256,212/256,199/256))\n",
    "    progression = []\n",
    "    pathways = []\n",
    "    expr = []\n",
    "    genes_used = []\n",
    "    trajs = []\n",
    "    \n",
    "    # Using k nearest points from the given start coordinate\n",
    "    start_points = getNNeighbors(start_coord, adata.obsm['X_pca'][:,0:2], k=num_traj)[0]\n",
    "    cell_phase_genes = get_cell_phase_genes(adata, None)\n",
    "    \n",
    "    # Obtains a trajectory --> cell phase score for each point in the trajectory\n",
    "    for i in range(num_traj):\n",
    "        path_sol, expr_output, all_genes, traj = get_LAP(adata.obsm['X_pca'][:,0:2][start_points[i]], end_coord, adata, num_points)\n",
    "        pathways.append(path_sol)\n",
    "        trajs.append(traj)\n",
    "        phase_scores = pd.DataFrame(batch_group_score(adata, expr_output, all_genes, cell_phase_genes))\n",
    "        phases = phase_scores.columns\n",
    "\n",
    "        gene_indexes = []\n",
    "        for j in range(len(genes)):\n",
    "            if genes[j] in adata.uns['LAP_pca']['genes'].tolist():\n",
    "                gene_indexes.append(all_genes.get_loc(genes[j]))\n",
    "                if i == 0:\n",
    "                    genes_used.append(genes[j])\n",
    "        \n",
    "        extracted_gene_expr = expr_output[:,gene_indexes]\n",
    "        expr.append(extracted_gene_expr)\n",
    "        \n",
    "        final_num = []\n",
    "        for i in range(len(phase_scores)):\n",
    "            line = np.array(phase_scores.loc[i])\n",
    "            line_max = np.max(line)\n",
    "            phase_index = np.where(line == line_max)[0][0]\n",
    "            final_num.append(phase_index)\n",
    "        progression.append(final_num)\n",
    "\n",
    "    pathways = np.array(pathways)\n",
    "    top_prog, top_paths, bottom_prog, bottom_paths, top_index, bottom_index = separate_paths2(progression, pathways)\n",
    "        \n",
    "    height = 0\n",
    "    for i in range(len(top_prog)):\n",
    "        x_val = 0\n",
    "        for j in range(len(top_prog[i])):\n",
    "            plt.plot([x_val, x_val + 1], [height,height], color=myColors[top_prog[i][j]], linewidth=35)\n",
    "            x_val += 1\n",
    "        height += 1\n",
    "\n",
    "    first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "    sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "    third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "    fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "    fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "    plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "    plt.ylabel(\"Cells\")\n",
    "    plt.xlabel(\"Top trajectories\")\n",
    "    plt.show()\n",
    "    \n",
    "    height = 0\n",
    "    for i in range(len(bottom_prog)):\n",
    "        x_val = 0\n",
    "        for j in range(len(bottom_prog[i])):\n",
    "            plt.plot([x_val, x_val + 1], [height,height], color=myColors[bottom_prog[i][j]], linewidth=60)\n",
    "            x_val += 1\n",
    "        height += 1\n",
    "\n",
    "    first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "    sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "    third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "    fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "    fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "    plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "    plt.ylabel(\"Cells\")\n",
    "    plt.xlabel(\"Bottom trajectories\")\n",
    "    plt.show()\n",
    "    \n",
    "    # Plots each trajectories cell phase score in a line plot\n",
    "#     height = 0\n",
    "#     for i in range(len(progression)):\n",
    "#         x_val = 0\n",
    "#         for j in range(len(progression[i])):\n",
    "#             plt.plot([x_val, x_val + 1], [height,height], color=myColors[progression[i][j]], linewidth=20)\n",
    "#             x_val += 1\n",
    "#         height += 1\n",
    "#     first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "#     sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "#     third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "#     fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "#     fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "#     plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "#     plt.ylabel(\"Cells\")\n",
    "#     plt.xlabel(\"Trajectory\")\n",
    "#     plt.show()\n",
    "    \n",
    "    # Plots each trajectory for reference\n",
    "#     for i in range(len(pathways)):\n",
    "#         plt.scatter(adata.obsm[\"X_pca\"][:,0], adata.obsm[\"X_pca\"][:,1], color = \"lightgrey\")\n",
    "#         plt.scatter(pathways[i][:,0], pathways[i][:,1])\n",
    "#         plt.show()\n",
    "#     plt.show()\n",
    "    \n",
    "    return top_prog, top_paths, bottom_prog, bottom_paths, expr, top_index, bottom_index, genes_used, trajs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ba637115",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Parameters: \n",
    "#     num_points: number of points for each trajectory\n",
    "#     num_traj: number of trajectories to generate\n",
    "# Returns:\n",
    "#     pathways: 2D list containing the coordinates for each trajectory\n",
    "#     progression: 2D list containing the cell cycle categories for each trajectory\n",
    "\n",
    "# Selects a bigger range of start points \n",
    "\n",
    "def cell_progression2(upper_left, lower_right, end_point, adata, num_points, num_traj, genes):\n",
    "    phases = np.array(['G1-S', 'S', 'G2-M', 'M', 'M-G1'], dtype=object)\n",
    "    myColors = ((255/256,91/256,165/256),(219/256,73/256,172/256), (176/256,124/256,218/256), (107/256,167/256,214/256),(100/256,212/256,199/256))\n",
    "    progression = []\n",
    "    pathways = []\n",
    "    expr = []\n",
    "    genes_used = []\n",
    "    trajs = []\n",
    "    \n",
    "    # Using k nearest points from the given start coordinate\n",
    "    start_points = []\n",
    "    x_num = None\n",
    "    y_num = None\n",
    "    for i in range(num_traj):\n",
    "        if upper_left[0] < lower_right[0]:\n",
    "            x_num = np.random.uniform(upper_left[0], lower_right[0])\n",
    "        else:\n",
    "            x_num = np.random.uniform( lower_right[0], upper_left[0])\n",
    "        \n",
    "        if upper_left[1] < lower_right[1]:\n",
    "            y_num = np.random.uniform(upper_left[1], lower_right[1])\n",
    "        else:\n",
    "            y_num = np.random.uniform( lower_right[1], upper_left[1])\n",
    "            \n",
    "        start_points.append([x_num, y_num])\n",
    "        \n",
    "    #start_points = getNNeighbors(start_coord, adata.obsm['X_pca'][:,0:2], k=num_traj)[0]\n",
    "    \n",
    "    cell_phase_genes = get_cell_phase_genes(adata, None)\n",
    "    \n",
    "    # Obtains a trajectory --> cell phase score for each point in the trajectory\n",
    "    for i in range(num_traj):\n",
    "        #path_sol, expr_output, all_genes = get_LAP(adata.obsm['X_pca'][:,0:2][start_points[i]], end_coord, adata, num_points)\n",
    "        path_sol, expr_output, all_genes, traj = get_LAP(start_points[i], end_coord, adata, num_points)\n",
    "        \n",
    "        #print(expr_output)\n",
    "        #print(start_points[i], end_coord)\n",
    "        #plt.scatter(adata.obsm['X_pca'][:,0], adata.obsm['X_pca'][:,1])\n",
    "        #plt.scatter(path_sol[:,0], path_sol[:,1])\n",
    "        #plt.show()\n",
    "        pathways.append(path_sol)\n",
    "        trajs.append(traj)\n",
    "        phase_scores = pd.DataFrame(batch_group_score(adata, expr_output, all_genes, cell_phase_genes))\n",
    "        phases = phase_scores.columns\n",
    "\n",
    "        gene_indexes = []\n",
    "        for j in range(len(genes)):\n",
    "            if genes[j] in adata.uns['LAP_pca']['genes'].tolist():\n",
    "                gene_indexes.append(all_genes.get_loc(genes[j]))\n",
    "                if i == 0:\n",
    "                    genes_used.append(genes[j])\n",
    "        \n",
    "        extracted_gene_expr = expr_output[:,gene_indexes]\n",
    "        expr.append(extracted_gene_expr)\n",
    "        \n",
    "        final_num = []\n",
    "        #display(phase_scores[0:10])\n",
    "        for i in range(len(phase_scores)):\n",
    "            line = np.array(phase_scores.loc[i])\n",
    "            line_max = np.max(line)\n",
    "            phase_index = np.where(line == line_max)[0][0]\n",
    "            final_num.append(phase_index)\n",
    "        progression.append(final_num)\n",
    "\n",
    "    pathways = np.array(pathways)\n",
    "    top_prog, top_paths, bottom_prog, bottom_paths, top_index, bottom_index = separate_paths2(progression, pathways)\n",
    "        \n",
    "    height = 0\n",
    "    for i in range(len(top_prog)):\n",
    "        x_val = 0\n",
    "        for j in range(len(top_prog[i])):\n",
    "            plt.plot([x_val, x_val + 1], [height,height], color=myColors[top_prog[i][j]], linewidth=35)\n",
    "            x_val += 1\n",
    "        height += 1\n",
    "\n",
    "    first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "    sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "    third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "    fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "    fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "    plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "    plt.ylabel(\"Cells\")\n",
    "    plt.xlabel(\"Top trajectories\")\n",
    "    plt.show()\n",
    "    \n",
    "    height = 0\n",
    "    for i in range(len(bottom_prog)):\n",
    "        x_val = 0\n",
    "        for j in range(len(bottom_prog[i])):\n",
    "            plt.plot([x_val, x_val + 1], [height,height], color=myColors[bottom_prog[i][j]], linewidth=60)\n",
    "            x_val += 1\n",
    "        height += 1\n",
    "\n",
    "    first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "    sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "    third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "    fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "    fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "    plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "    plt.ylabel(\"Cells\")\n",
    "    plt.xlabel(\"Bottom trajectories\")\n",
    "    plt.show()\n",
    "    \n",
    "    # Plots each trajectories cell phase score in a line plot\n",
    "#     height = 0\n",
    "#     for i in range(len(progression)):\n",
    "#         x_val = 0\n",
    "#         for j in range(len(progression[i])):\n",
    "#             plt.plot([x_val, x_val + 1], [height,height], color=myColors[progression[i][j]], linewidth=20)\n",
    "#             x_val += 1\n",
    "#         height += 1\n",
    "#     first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "#     sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "#     third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "#     fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "#     fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "#     plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "#     plt.ylabel(\"Cells\")\n",
    "#     plt.xlabel(\"Trajectory\")\n",
    "#     plt.show()\n",
    "    \n",
    "    # Plots each trajectory for reference\n",
    "#     for i in range(len(pathways)):\n",
    "#         plt.scatter(adata.obsm[\"X_pca\"][:,0], adata.obsm[\"X_pca\"][:,1], color = \"lightgrey\")\n",
    "#         plt.scatter(pathways[i][:,0], pathways[i][:,1])\n",
    "#         plt.show()\n",
    "#     plt.show()\n",
    "    \n",
    "    return top_prog, top_paths, bottom_prog, bottom_paths, expr, top_index, bottom_index, genes_used, trajs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "128d0367",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Parameters: \n",
    "#     num_points: number of points for each trajectory\n",
    "#     num_traj: number of trajectories to generate\n",
    "# Returns:\n",
    "#     pathways: 2D list containing the coordinates for each trajectory\n",
    "#     progression: 2D list containing the cell cycle categories for each trajectory\n",
    "\n",
    "# Selects a bigger range of start points and end points\n",
    "\n",
    "def cell_progression3(upper_left, lower_right, upper_left2, lower_right2, adata, num_points, num_traj, genes):\n",
    "    phases = np.array(['G1-S', 'S', 'G2-M', 'M', 'M-G1'], dtype=object)\n",
    "    myColors = ((255/256,91/256,165/256),(219/256,73/256,172/256), (176/256,124/256,218/256), (107/256,167/256,214/256),(100/256,212/256,199/256))\n",
    "    progression = []\n",
    "    pathways = []\n",
    "    expr = []\n",
    "    genes_used = []\n",
    "    trajs = []\n",
    "    \n",
    "    # Using k nearest points from the given start coordinate\n",
    "    start_points = []\n",
    "    end_points = []\n",
    "    x_num = None\n",
    "    y_num = None\n",
    "    x_num2 = None\n",
    "    y_num2 = None\n",
    "    for i in range(num_traj):\n",
    "        if upper_left[0] < lower_right[0]:\n",
    "            x_num = np.random.uniform(upper_left[0], lower_right[0])\n",
    "        else:\n",
    "            x_num = np.random.uniform( lower_right[0], upper_left[0])\n",
    "        \n",
    "        if upper_left[1] < lower_right[1]:\n",
    "            y_num = np.random.uniform(upper_left[1], lower_right[1])\n",
    "        else:\n",
    "            y_num = np.random.uniform( lower_right[1], upper_left[1])\n",
    "        \n",
    "        if upper_left2[0] < lower_right2[0]:\n",
    "            x_num2 = np.random.uniform(upper_left2[0], lower_right2[0])\n",
    "        else:\n",
    "            x_num2 = np.random.uniform( lower_right2[0], upper_left2[0])\n",
    "        \n",
    "        if upper_left2[1] < lower_right2[1]:\n",
    "            y_num2 = np.random.uniform(upper_left2[1], lower_right2[1])\n",
    "        else:\n",
    "            y_num2 = np.random.uniform( lower_right2[1], upper_left2[1])\n",
    "        \n",
    "        start_points.append([x_num, y_num])\n",
    "        end_points.append([x_num2, y_num2])\n",
    "        \n",
    "        \n",
    "    #start_points = getNNeighbors(start_coord, adata.obsm['X_pca'][:,0:2], k=num_traj)[0]\n",
    "    \n",
    "    cell_phase_genes = get_cell_phase_genes(adata, None)\n",
    "    \n",
    "    # Obtains a trajectory --> cell phase score for each point in the trajectory\n",
    "    for i in range(num_traj):\n",
    "        #path_sol, expr_output, all_genes = get_LAP(adata.obsm['X_pca'][:,0:2][start_points[i]], end_coord, adata, num_points)\n",
    "        path_sol, expr_output, all_genes, traj = get_LAP(start_points[i], end_points[i], adata, num_points)\n",
    "        pathways.append(path_sol)\n",
    "        trajs.append(traj)\n",
    "        phase_scores = pd.DataFrame(batch_group_score(adata, expr_output, all_genes, cell_phase_genes))\n",
    "        phases = phase_scores.columns\n",
    "\n",
    "        gene_indexes = []\n",
    "        for j in range(len(genes)):\n",
    "            if genes[j] in adata.uns['LAP_pca']['genes'].tolist():\n",
    "                gene_indexes.append(all_genes.get_loc(genes[j]))\n",
    "                if i == 0:\n",
    "                    genes_used.append(genes[j])\n",
    "        \n",
    "        extracted_gene_expr = expr_output[:,gene_indexes]\n",
    "        expr.append(extracted_gene_expr)\n",
    "        \n",
    "        final_num = []\n",
    "        #display(phase_scores)\n",
    "        for i in range(len(phase_scores)):\n",
    "            line = np.array(phase_scores.loc[i])\n",
    "            line_max = np.max(line)\n",
    "            #print(line, line_max)\n",
    "            phase_index = np.where(line == line_max)[0][0]\n",
    "            final_num.append(phase_index)\n",
    "        progression.append(final_num)\n",
    "\n",
    "    pathways = np.array(pathways)\n",
    "    top_prog, top_paths, bottom_prog, bottom_paths, top_index, bottom_index = separate_paths2(progression, pathways)\n",
    "        \n",
    "    height = 0\n",
    "    for i in range(len(top_prog)):\n",
    "        x_val = 0\n",
    "        for j in range(len(top_prog[i])):\n",
    "            plt.plot([x_val, x_val + 1], [height,height], color=myColors[top_prog[i][j]], linewidth=35)\n",
    "            x_val += 1\n",
    "        height += 1\n",
    "\n",
    "    first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "    sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "    third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "    fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "    fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "    plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "    plt.ylabel(\"Cells\")\n",
    "    plt.xlabel(\"Top trajectories\")\n",
    "    plt.show()\n",
    "    \n",
    "    height = 0\n",
    "    for i in range(len(bottom_prog)):\n",
    "        x_val = 0\n",
    "        for j in range(len(bottom_prog[i])):\n",
    "            plt.plot([x_val, x_val + 1], [height,height], color=myColors[bottom_prog[i][j]], linewidth=60)\n",
    "            x_val += 1\n",
    "        height += 1\n",
    "\n",
    "    first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "    sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "    third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "    fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "    fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "    plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "    plt.ylabel(\"Cells\")\n",
    "    plt.xlabel(\"Bottom trajectories\")\n",
    "    plt.show()\n",
    "    \n",
    "    # Plots each trajectories cell phase score in a line plot\n",
    "#     height = 0\n",
    "#     for i in range(len(progression)):\n",
    "#         x_val = 0\n",
    "#         for j in range(len(progression[i])):\n",
    "#             plt.plot([x_val, x_val + 1], [height,height], color=myColors[progression[i][j]], linewidth=20)\n",
    "#             x_val += 1\n",
    "#         height += 1\n",
    "#     first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "#     sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "#     third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "#     fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "#     fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "#     plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "#     plt.ylabel(\"Cells\")\n",
    "#     plt.xlabel(\"Trajectory\")\n",
    "#     plt.show()\n",
    "    \n",
    "    # Plots each trajectory for reference\n",
    "#     for i in range(len(pathways)):\n",
    "#         plt.scatter(adata.obsm[\"X_pca\"][:,0], adata.obsm[\"X_pca\"][:,1], color = \"lightgrey\")\n",
    "#         plt.scatter(pathways[i][:,0], pathways[i][:,1])\n",
    "#         plt.show()\n",
    "#     plt.show()\n",
    "    \n",
    "    return top_prog, top_paths, bottom_prog, bottom_paths, expr, top_index, bottom_index, genes_used, trajs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9a4f3ae",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "accc8901",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a94822ff",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# def plot_trajectory(start_coord, end_coord, adata, number_of_cells):\n",
    "    \n",
    "#     start_indices = getNNeighbors(start_coord, adata.obsm['X_umap'], k=number_of_cells)[0]\n",
    "#     #end_indices = getNNeighbors([8.5,1], adata.obsm['X_umap'], k=10)[0]\n",
    "    \n",
    "#     for i in range(len(start_indices)):\n",
    "#         start = adata.obsm[\"X_umap\"][start_indices[i], :]\n",
    "#         print(\"Starting point: \", start)\n",
    "#         print(\"End point: \", end_coord)\n",
    "#         path_sol = dyn.pd.least_action(adata, start, end_coord)\n",
    "#         print(len(path_sol))\n",
    "\n",
    "#         plt.scatter(adata.obsm[\"X_umap\"][:,0], adata.obsm[\"X_umap\"][:,1], color = \"lightgrey\")\n",
    "#         plt.scatter(path_sol[:,0], path_sol[:,1], c=path_sol[:,0])\n",
    "#         plt.scatter(adata.obsm[\"X_umap\"][start_indices[i],0], adata.obsm[\"X_umap\"][start_indices[i],1], marker = \"*\", color = \"red\", label=\"start\")\n",
    "#         #plt.scatter(adata.obsm[\"X_umap\"][end_coord,0], adata.obsm[\"X_umap\"][end_coord,1], marker = \"*\", color = \"black\", label = \"end\")\n",
    "#         plt.scatter(end_coord[0], end_coord[1], marker = \"*\", color = \"black\", label = \"end\")\n",
    "#         #plt.title(\"TGFB dose 12.5 - least action path\")\n",
    "#         plt.legend(loc=\"upper left\")\n",
    "#         filename = \"Plots/EMT_trajectory/Trajectories/traj_\"+str(i)\n",
    "#         plt.savefig(filename, dpi=300)\n",
    "#         plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f33fed61",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def finding_gene_names(gene_list, adata):\n",
    "    indexes = []\n",
    "    gene_names = []\n",
    "    for i in range(len(gene_list)):\n",
    "        gene_index = np.where(adata.var['var_names'] == gene_list[i])[0]\n",
    "        if gene_index.shape[0] > 0:\n",
    "            indexes.append(gene_index[0])\n",
    "            gene_names.append(gene_list[i])\n",
    "    gene_names = np.array(gene_names)\n",
    "    return gene_names, indexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a1816060",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## BUGS ##\n",
    "def get_gene_progression(path_sol, adata, gene_names, indexes, groups):\n",
    "    gene_progression = []\n",
    "    interval = 1/groups\n",
    "    \n",
    "    gene_max = np.max(adata.X.toarray)\n",
    "    gene_min = np.min(adata.X.toarray)\n",
    "    \n",
    "    for i in range(0,len(path_sol),1):\n",
    "        indices = getNNeighbors(path_sol[i], adata.obsm['X_umap'], k=100)[0]\n",
    "\n",
    "        temp = adata[indices,indexes]\n",
    "        gene_exp = temp.X.toarray()\n",
    "\n",
    "        mean = np.mean(gene_exp, axis = 0)\n",
    "        order = np.argsort(mean)\n",
    "        gene_exp2 = gene_exp[:, order]\n",
    "\n",
    "        #num = np.arange(0,len(mean))\n",
    "        \n",
    "        total = interval\n",
    "        groupings = []\n",
    "        while(total<=1):\n",
    "            if total == 1:\n",
    "                groupings.append(np.where(mean >= np.quantile(mean, total - interval))[0].tolist())\n",
    "            else:\n",
    "                groupings.append(np.where((mean < np.quantile(mean,total)) & (mean >= np.quantile(mean, total - interval)))[0].tolist())\n",
    "            total = total + interval\n",
    "        temp2 = []\n",
    "        for j in range(len(groupings)):\n",
    "            temp2.append(gene_names[groupings[j]].tolist())\n",
    "        gene_progression.append(temp2)\n",
    "        \n",
    "        #group1 = np.where(mean < np.quantile(mean, 0.33))[0]\n",
    "        #group3 = np.where(mean > np.quantile(mean, 0.66))[0]\n",
    "        #group2 = list(set(num.tolist()) - set(group1.tolist()))\n",
    "        #group2 = list(set(group2) - set(group3.tolist()))\n",
    "\n",
    "        #gene_progression.append([gene_names[group1].tolist(), gene_names[group2].tolist(), gene_names[group3].tolist()])\n",
    "\n",
    "        #ax = sns.heatmap(gene_exp2, xticklabels=gene_names[order])\n",
    "        #plt.yticks(rotation=360)\n",
    "        #plt.xticks(rotation=45)\n",
    "        #plt.savefig(\"Plots/EMT_trajectory/1000_genes/trajectory_\"+str(i), dpi=300)\n",
    "        #plt.show()\n",
    "    return gene_progression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "21d0af02",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def get_gene_progression2(path_sol, adata, gene_names, indexes, groups):\n",
    "    #Looping through all the genes\n",
    "    genes_progression = []\n",
    "    genes_quantiles = []\n",
    "    for i in range(gene_names.shape[0]):\n",
    "        #print(gene_names[i])\n",
    "        exp = []\n",
    "        #Looping through cells along the trajectory path\n",
    "        for j in range(len(path_sol)):\n",
    "            indices = getNNeighbors(path_sol[j], adata.obsm['X_umap'], k=20)[0]\n",
    "            temp = adata[indices,indexes[i]]\n",
    "            gene_exp = temp.X.toarray()\n",
    "            exp.append(gene_exp)\n",
    "        interval = 1/groups\n",
    "        quantiles = np.arange(0, 1.01, interval)\n",
    "        #for k in range(len(quantiles)):\n",
    "        #    print(round(quantiles[k],3), \"\\t\", np.quantile(exp, quantiles[k]))\n",
    "        #print(quantiles)\n",
    "        #print(i, \":\", len(exp))\n",
    "        if np.quantile(exp, quantiles[groups]) == 0:\n",
    "            genes_progression.append(np.repeat([0], len(exp)))\n",
    "            genes_quantiles.append(np.quantile(exp, quantiles))\n",
    "        else:\n",
    "            gene_prog = []\n",
    "            for j in range(len(path_sol)):\n",
    "                indices = getNNeighbors(path_sol[j], adata.obsm['X_umap'], k=20)[0]\n",
    "                temp = adata[indices,indexes[i]]\n",
    "                exp_mean = np.mean(temp.X.toarray())\n",
    "                for k in range(1,len(quantiles)):\n",
    "                    if (exp_mean >= np.quantile(exp, quantiles[k-1])) & (exp_mean < np.quantile(exp, quantiles[k])):\n",
    "                        gene_prog.append(k-1)\n",
    "            genes_progression.append(gene_prog)\n",
    "            genes_quantiles.append(np.quantile(exp, quantiles))\n",
    "    print(len(genes_progression))\n",
    "    print(len(genes_quantiles))\n",
    "    return genes_progression, genes_quantiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "305da7c5",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_gene_progression2(gene_names, progression, gene_quantiles, groups, filename, save=True):\n",
    "    timepoints = np.arange(len(progression[0]))\n",
    "    for i in range(len(progression)):\n",
    "        #plt.plot(timepoints, progression[i], marker = 'o')\n",
    "        #z = np.polyfit(timepoints, progression[i], 2)\n",
    "        #p = np.poly1d(z)\n",
    "        print(len(progression[i]))\n",
    "        plt.bar(timepoints, progression[i], color=sns.color_palette(\"rocket\", len(timepoints)), edgecolor='black', width=0.7)\n",
    "        #plt.plot(timepoints, p(timepoints), 'r--')\n",
    "        \n",
    "        plt.title(gene_names[i])\n",
    "        plt.ylim(-0.5, groups+0.5-1)\n",
    "        #file = \"Plots/EMT_trajectory/\" + filename + \"/\" + gene_names[i] + \"_bar_progression\"\n",
    "        file = filename + \"/\" + gene_names[i] + \"_bar_progression\"\n",
    "        #print(gene_quantiles[i])\n",
    "        plt.xlabel(\"Group 0: \" + str(round(gene_quantiles[i][0],2)) + \"   -   \" + \"Group \" + str(groups) + \": \" + str(round(gene_quantiles[i][groups],2)))\n",
    "        if save == True:\n",
    "            plt.savefig(file, dpi=300)\n",
    "        plt.tight_layout()\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "4721a7d4",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def get_ge_along_traj(gene_names, path, adata, groups, directory, save):\n",
    "    parent_dir = \"/Users/sophiahu/Documents/Xing lab/MCF10A_data_NEW/notebooks/vs_notebooks/Plots/EMT_trajectory/\"\n",
    "    new_path = os.path.join(parent_dir, directory)\n",
    "    if os.path.isdir(new_path) == False:\n",
    "        os.mkdir(new_path)\n",
    "    gene_names, indexes = finding_gene_names(gene_names, adata)\n",
    "    gene_prog, gene_quantiles = get_gene_progression2(path, adata, gene_names, indexes, groups)\n",
    "    #print(len(gene_prog))\n",
    "    #print(len(gene_quantiles))\n",
    "    plot_gene_progression2(gene_names, gene_prog, gene_quantiles, groups, new_path, save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "b329a32b",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_acceleration_traj(gene_names, indexes, path, adata):\n",
    "    for i in range(len(gene_names)):\n",
    "        accel_means = []\n",
    "        for j in range(len(path_sol)):\n",
    "            indices = getNNeighbors(path_sol[j], adata.obsm['X_umap'], k=20)[0]\n",
    "            temp = adata.layers['acceleration'][indices, indexes[i]]\n",
    "            accel_means.append(np.mean(temp))\n",
    "        x_axis = np.arange(0, len(accel_means))\n",
    "        plt.plot(x_axis, accel_means)\n",
    "        plt.title(gene_names[i])\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3f1d7fb6",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_cell_cycle_along_traj(path, adata):\n",
    "    phases = np.array(['M-G1','G1-S', 'S','G2-M','M'], dtype=object)\n",
    "    progression = []\n",
    "    for i in range(len(path)):\n",
    "        indices = getNNeighbors(path[i], adata.obsm['X_umap'], k=20)[0]\n",
    "        temp = adata[indices, :]\n",
    "        phase_tracker = []\n",
    "        for j in range(phases.shape[0]):\n",
    "            phase_tracker.append(len(np.where(temp.obs['cell_cycle_phase'] == phases[j])[0]))\n",
    "        max_phase = max(phase_tracker)\n",
    "        phase = np.where(np.array(phase_tracker) == max_phase)[0][0]\n",
    "        progression.append(phase)\n",
    "\n",
    "    phases = phases.tolist()\n",
    "    plt.bar(np.arange(0,len(path),1).tolist(), progression)\n",
    "    plt.ylim(0,4.5)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e2f1da26",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def get_cell_cycle_along_traj(path, adata):\n",
    "    phases = np.array(['M-G1','G1-S', 'S','G2-M','M'], dtype=object)\n",
    "    progression = []\n",
    "    for i in range(path.shape[0]):\n",
    "        indices = getNNeighbors(path[i], adata.obsm['X_umap'], k=20)[0]\n",
    "        temp = adata[indices, :]\n",
    "        phase_tracker = []\n",
    "        for j in range(phases.shape[0]):\n",
    "            phase_tracker.append(len(np.where(temp.obs['cell_cycle_phase'] == phases[j])[0]))\n",
    "        max_phase = max(phase_tracker)\n",
    "        phase = np.where(np.array(phase_tracker) == max_phase)[0][0]\n",
    "        progression.append(phase)\n",
    "    return progression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "3fe7cee9",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Getting cell cycle phase for a single cell instead of averaging it\n",
    "def get_cell_cycle_along_traj2(path, adata):\n",
    "    phases = np.array(['M-G1','G1-S', 'S','G2-M','M'], dtype=object)\n",
    "    progression = []\n",
    "    for i in range(len(path)):\n",
    "        index = getNNeighbors(path[i], adata.obsm['X_umap'], k=1)[0]\n",
    "        temp = adata[index, :]\n",
    "        phase = np.where(phases == temp.obs['cell_cycle_phase'][0])[0][0]\n",
    "        progression.append(phase)\n",
    "    return progression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1e8194b",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "427a2c8b",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def multiple_traj_cell_cycle_progression(start_coord, end_coord, adata, cells):\n",
    "    indices = getNNeighbors(start_coord, adata.obsm['X_umap'], k=cells)[0] \n",
    "    cell_progressions = []\n",
    "    paths = []\n",
    "    for i in range(len(indices)):\n",
    "        path_sol = dyn.pd.least_action(adata, adata.obsm['X_umap'][indices[i],:], end_coord)\n",
    "        paths.append(path_sol)\n",
    "        to_add = get_cell_cycle_along_traj(path_sol, adata)\n",
    "        print(len(to_add))\n",
    "        cell_progressions.append(to_add)\n",
    "    return cell_progressions, np.array(paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "3cdbe513",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Plot trajectories with very different start points \n",
    "def multiple_traj_cell_cycle_progression2(top_left_corner, bottom_right_corner, end_coord, adata, cells):\n",
    "    cell_progressions = []\n",
    "    paths = []\n",
    "    for i in range(cells):\n",
    "        x_val = np.random.uniform(top_left_corner[0], bottom_right_corner[0])\n",
    "        y_val = np.random.uniform(bottom_right_corner[1], top_left_corner[1])\n",
    "\n",
    "        index = getNNeighbors([x_val, y_val], adata.obsm['X_umap'], k=1)[0]\n",
    "        path_sol = dyn.pd.least_action(adata, adata.obsm['X_umap'][index,:][0], end_coord)\n",
    "        paths.append(path_sol)\n",
    "        to_add = get_cell_cycle_along_traj(path_sol, adata)\n",
    "        print(len(to_add))\n",
    "        cell_progressions.append(to_add)\n",
    "    return cell_progressions, np.array(paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "ecd883e4",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Plot trajectories with very different start points \n",
    "def multiple_traj_cell_cycle_progression3(top_left_corner, bottom_right_corner, end_coord, adata, cells):\n",
    "    cell_progressions = []\n",
    "    paths = []\n",
    "    count = 0\n",
    "    #for i in range(cells):\n",
    "    while count < cells:\n",
    "        x_val = np.random.uniform(top_left_corner[0], bottom_right_corner[0])\n",
    "        y_val = np.random.uniform(bottom_right_corner[1], top_left_corner[1])\n",
    "        index = getNNeighbors([x_val, y_val], adata.obsm['X_umap'], k=1)[0]\n",
    "        path_sol = dyn.pd.least_action(adata, adata.obsm['X_umap'][index,:][0], end_coord)\n",
    "        toAdd = True\n",
    "        for j in range(len(paths)):\n",
    "            dtw_distance, _ = fastdtw(path_sol[:,0], path_sol[:,1], dist=euclidean)\n",
    "            if dtw_distance < 15:\n",
    "                toAdd = False\n",
    "        if toAdd == True:\n",
    "            paths.append(path_sol)\n",
    "            to_add = get_cell_cycle_along_traj(path_sol, adata)\n",
    "            print(len(to_add))\n",
    "            cell_progressions.append(to_add)\n",
    "            count += 1\n",
    "    return cell_progressions, np.array(paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39600367",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0273a747",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "abc9f9bb",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Separates the trajectories as either coming from the top or bottom\n",
    "def separate_paths(cell_prog, paths):\n",
    "    top_paths = []\n",
    "    bottom_paths = []\n",
    "    top_path_cell_progressions = []\n",
    "    bottom_path_cell_progressions = []\n",
    "    top_index = []\n",
    "    bottom_index = []\n",
    "    \n",
    "    for i in range(len(paths)):\n",
    "        start_y = paths[i][0,1]\n",
    "        length = paths[i].shape[0]\n",
    "\n",
    "        #if np.where(paths[i][:,1] < start_y)[0].shape[0] > 3*(paths[i].shape[0]/4):\n",
    "        if np.where(paths[i][int(1.5*(length/4)):int(3*(length/4)),1] < start_y)[0].shape[0] > 0:\n",
    "            bottom_paths.append(paths[i])\n",
    "            bottom_path_cell_progressions.append(cell_prog[i])\n",
    "            bottom_index.append(i)\n",
    "        else:\n",
    "            top_paths.append(paths[i])\n",
    "            top_path_cell_progressions.append(cell_prog[i])\n",
    "            top_index.append(i)\n",
    "            \n",
    "    top_paths = np.array(top_paths)\n",
    "    bottom_paths = np.array(bottom_paths)\n",
    "    top_path_cell_progressions = np.array(top_path_cell_progressions)\n",
    "    bottom_path_cell_progressions = np.array(bottom_path_cell_progressions)\n",
    "    \n",
    "    print(top_paths.shape[0])\n",
    "    print(bottom_paths.shape[0])\n",
    "    \n",
    "    return top_path_cell_progressions, top_paths, bottom_path_cell_progressions, bottom_paths, top_index, bottom_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d710f35",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Using the slope to determine if points are above or below the calculated slope\n",
    "# Determines if a path is going along the top or bottom path\n",
    "def separate_paths2(cell_prog, paths):\n",
    "    top_paths = []\n",
    "    bottom_paths = []\n",
    "    top_path_cell_progressions = []\n",
    "    bottom_path_cell_progressions = []\n",
    "    top_index = []\n",
    "    bottom_index = []\n",
    "    \n",
    "    for i in range(len(paths)):\n",
    "        start_y = paths[i][0,1]\n",
    "        start_x = paths[i][0,0]\n",
    "        end_y = paths[i][len(paths[i])-1,1]\n",
    "        end_x = paths[i][len(paths[i])-1,0]\n",
    "        \n",
    "        slope = (end_y - start_y)/(end_x - start_x)\n",
    "        b = start_y - slope*start_x\n",
    "        \n",
    "        top = 0\n",
    "        bottom = 0\n",
    "        for j in range(len(paths[i])):\n",
    "            y = paths[i][j,1]\n",
    "            x = paths[i][j,0]\n",
    "            if y > slope*x + b:\n",
    "                top +=1\n",
    "            elif y < slope*x + b:\n",
    "                bottom += 1\n",
    "        \n",
    "        if top > bottom:\n",
    "            top_paths.append(paths[i])\n",
    "            top_path_cell_progressions.append(cell_prog[i])\n",
    "            top_index.append(i)       \n",
    "        else:\n",
    "            bottom_paths.append(paths[i])\n",
    "            bottom_path_cell_progressions.append(cell_prog[i])\n",
    "            bottom_index.append(i)\n",
    "            \n",
    "    top_paths = np.array(top_paths)\n",
    "    bottom_paths = np.array(bottom_paths)\n",
    "    top_path_cell_progressions = np.array(top_path_cell_progressions)\n",
    "    bottom_path_cell_progressions = np.array(bottom_path_cell_progressions)\n",
    "    \n",
    "    print(top_paths.shape[0])\n",
    "    print(bottom_paths.shape[0])\n",
    "    \n",
    "    return top_path_cell_progressions, top_paths, bottom_path_cell_progressions, bottom_paths, top_index, bottom_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "8e99003e",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_cell_progression_lineplot(cell_progression, paths, intermediate_coord, line_width=10):\n",
    "    interval_cell_progression = []\n",
    "    cell_progression = np.array(cell_progression)\n",
    "    path_len = []\n",
    "    \n",
    "    for j in range(paths.shape[0]):\n",
    "#         x_coords = np.linspace(paths[j][0,0], paths[j][paths[j].shape[0]-1,0], paths[j].shape[0])\n",
    "#         coords = np.vstack((paths[j][:,0], np.repeat(0, paths[j].shape[0]))).T\n",
    "#         interval_coords = np.vstack((x_coords, np.repeat(0, x_coords.shape[0]))).T\n",
    "\n",
    "#         temp = []\n",
    "#         for i in range(interval_coords.shape[0]): \n",
    "#             index = getNNeighbors(interval_coords[i], coords, 1)[0][0]\n",
    "\n",
    "#             temp.append(cell_progression[j][index])\n",
    "#         interval_cell_progression.append(temp)\n",
    "        path_len.append(paths[j].shape[0])\n",
    "\n",
    "    # Plotting using multiple line plots\n",
    "    import matplotlib.patches as mpatches\n",
    "\n",
    "    myColors = ((248/256,177/256,149/256),(246/256,114/256,128/256),(192/256,108/256,132/256),(108/256,91/256,123/256),(53/256,92/256,125/256))\n",
    "\n",
    "    first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "    sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "    third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "    fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "    fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "    plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "    ax = plt.axes()\n",
    "    ax.set_facecolor(\"white\")\n",
    "\n",
    "#     cell = 0\n",
    "#     for i in range(len(interval_cell_progression)):\n",
    "#         for j in range(len(interval_cell_progression[i])):\n",
    "#             plt.plot([j, j+1], [cell, cell], color = myColors[interval_cell_progression[i][j]], linewidth=line_width)\n",
    "#         cell = cell + 1\n",
    "#     #plt.ylabel(\"Cells\")\n",
    "#     plt.xlabel(\"Trajectory\")\n",
    "#     plt.yticks([])\n",
    "#     plt.xlim(-1, max(path_len)+1)\n",
    "\n",
    "    cell = 0\n",
    "    for i in range(len(cell_progression)):\n",
    "        for j in range(len(cell_progression[i])):\n",
    "            plt.plot([j, j+1], [cell, cell], color = myColors[cell_progression[i][j]], linewidth=line_width)\n",
    "        cell = cell + 1\n",
    "    #plt.ylabel(\"Cells\")\n",
    "    plt.xlabel(\"Trajectory\")\n",
    "    plt.yticks([])\n",
    "    plt.xlim(-1, max(path_len)+1)\n",
    "\n",
    "    cell_range = paths[0][paths[0].shape[0]-1,0] - paths[0][0,0]\n",
    "    interval = cell_range/4\n",
    "    cells3 = np.where((paths[0][:,0] > intermediate_coord[0]-interval)&(paths[0][:,0] < intermediate_coord[0]+interval))[0]\n",
    "    \n",
    "    #bounds = [0, np.min(cells3), np.max(cells3), paths[0].shape[0]-1]\n",
    "    bounds = [0,20, 40, max(path_len)-1]\n",
    "    #print(bounds)\n",
    "    #norm = mpl.colors.Normalize(vmin=0,vmax=2)\n",
    "\n",
    "#     n_clusters = 3\n",
    "#     cmap = plt.get_cmap('Set3', n_clusters)\n",
    "#     norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "#     sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "#     #sm.set_array([])\n",
    "\n",
    "#     cbar = plt.colorbar(sm,\n",
    "#                  orientation=\"horizontal\")\n",
    "\n",
    "#     #tick_locs = (np.arange(3) + 0.5)*(3-1)/3\n",
    "#     tick_locs = [(0+bounds[1]/2, (bounds[1]+bounds[2]/2), (bounds[2]+bounds[3])/2)]\n",
    "#     cbar.set_ticks(tick_locs)\n",
    "#     cbar.set_ticklabels(['E','I','M'])\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(6, 1))\n",
    "    fig.subplots_adjust(bottom=0.6)\n",
    "    \n",
    "    cmap = plt.get_cmap('Set3', 3)\n",
    "\n",
    "    norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "    #position=fig1.add_axes([0.93,0.1,0.02,0.35]) \n",
    "    \n",
    "    cbar = fig.colorbar(\n",
    "        mpl.cm.ScalarMappable(cmap=cmap, norm=norm),\n",
    "        cax=ax,\n",
    "        #cax=[0.93,0.1,0.02,0.35],\n",
    "        spacing='proportional',\n",
    "        orientation='horizontal')\n",
    "    \n",
    "    tick_locs = [(0+bounds[1]/2, (bounds[1]+bounds[2])/2, (bounds[2]+bounds[3])/2)]\n",
    "    cbar.set_ticks(tick_locs)\n",
    "    cbar.set_ticklabels(['Epithelial','Intermediate','Mesenchymal'])\n",
    "\n",
    "    \n",
    "    \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "1b48cee4",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_cell_progression_lineplot2(cell_progression, paths, intermediate_coord, line_width=10):\n",
    "    interval_cell_progression = []\n",
    "    cell_progression = np.array(cell_progression)\n",
    "    path_len = []\n",
    "    intermediate_len = []\n",
    "    index = 0\n",
    "    \n",
    "    n_clusters = 5\n",
    "    #cmap = plt.get_cmap('Plasma', n_clusters)\n",
    "    cmap = plt.cm.get_cmap('Purples')\n",
    "    \n",
    "    #myColors = ((248/256,177/256,149/256),(246/256,114/256,128/256),(192/256,108/256,132/256),(108/256,91/256,123/256),(53/256,92/256,125/256))\n",
    "    myColors = ((255/256,91/256,165/256),(219/256,73/256,172/256), (176/256,124/256,218/256), (107/256,167/256,214/256),(100/256,212/256,199/256))\n",
    "    \n",
    "    for i in range(paths.shape[0]):\n",
    "        cell_range = paths[i][paths[i].shape[0]-1,0] - paths[i][0,0]\n",
    "        interval = cell_range/3\n",
    "        cells = np.where((paths[i][:,0] > intermediate_coord[0]-interval)&(paths[i][:,0] < intermediate_coord[0]+interval))[0]\n",
    "        intermediate_len.append(cells.shape[0])\n",
    "        path_len.append(paths[i].shape[0])\n",
    "        \n",
    "    min_path = min(path_len)\n",
    "    min_intermediate = min(intermediate_len)\n",
    "    print(\"Min length: \", min_path)\n",
    "    print(\"Min intermediate: \", min_intermediate)\n",
    "        \n",
    "    for i in range(paths.shape[0]):    \n",
    "        cell_range = paths[i][paths[i].shape[0]-1,0] - paths[i][0,0]\n",
    "        interval = cell_range/3\n",
    "\n",
    "        cells = np.where(paths[i][:,0] < intermediate_coord[0]-interval)[0]\n",
    "        cells2 = np.where(paths[i][:,0] > intermediate_coord[0]+interval)[0]\n",
    "        cells3 = np.where((paths[i][:,0] > intermediate_coord[0]-interval)&(paths[i][:,0] < intermediate_coord[0]+interval))[0]\n",
    "        #outside_cells = np.concatenate((cells, cells2))\n",
    "        \n",
    "        print(\"start:\", cells.shape[0])\n",
    "        print(\"end:\", cells2.shape[0])\n",
    "        print(\"middle:\", cells3.shape[0])\n",
    "        \n",
    "        #if cells3.shape[0] > min_intermediate:\n",
    "        #    cells3 = random.sample(cells3.tolist(), min_intermediate)\n",
    "        #    cells3 = np.array(cells3)\n",
    "        \n",
    "        for j in range(cells.shape[0]):\n",
    "            #plt.plot([j-cells.shape[0], j+1-cells.shape[0]], [index, index], color = myColors[cell_progression[i][cells[j]]], linewidth=line_width )\n",
    "            plt.plot([j-cells.shape[0], j+1-cells.shape[0]], [index, index], color = cmap(cell_progression[i][cells[j]]/phases.shape[0]), linewidth=line_width )\n",
    "        \n",
    "        for j in range(cells3.shape[0]):\n",
    "            #plt.plot([j, j+1], [index, index], color = myColors[cell_progression[i][cells3[j]]], linewidth=line_width )\n",
    "            plt.plot([j, j+1], [index, index], color = cmap(cell_progression[i][cells3[j]]/phases.shape[0]), linewidth=line_width )\n",
    "            \n",
    "        for j in range(cells2.shape[0]):\n",
    "            #plt.plot([j+cells3.shape[0], j+1+cells3.shape[0]], [index, index], color = myColors[cell_progression[i][cells2[j]]], linewidth=line_width )\n",
    "            plt.plot([j+cells3.shape[0], j+1+cells3.shape[0]], [index, index], color = cmap(cell_progression[i][cells2[j]]/phases.shape[0]), linewidth=line_width )\n",
    "            \n",
    "        index = index + 1\n",
    "        \n",
    "    plt.vlines(0,ymin=0, ymax=paths.shape[0]-0.5, linestyles='dashed')\n",
    "    plt.vlines(min_intermediate,ymin=0, ymax=paths.shape[0]-0.5, linestyles='dashed')\n",
    "        \n",
    "#         outside_num = min_path - cells3.shape[0]\n",
    "        \n",
    "#         num_to_choose = int(outside_num/2)\n",
    "#         beginning = []\n",
    "#         ending = []\n",
    "        \n",
    "#         if cells.shape[0] < num_to_choose:\n",
    "#             beginning = cells.tolist()\n",
    "#             diff = num_to_choose - len(beginning)\n",
    "#             ending = random.sample(cells2.tolist(), num_to_choose + diff)\n",
    "            \n",
    "#         if cells2.shape[0] < num_to_choose:\n",
    "#             ending = cells2.tolist()\n",
    "#             diff = num_to_choose - len(ending)\n",
    "#             beginning = random.sample(cells.tolist(), num_to_choose + diff) \n",
    "            \n",
    "#         if cells.shape[0] > num_to_choose & cells2.shape[0] > num_to_choose:\n",
    "#             beginning = random.sample(cells.tolist(), num_to_choose) \n",
    "#             ending = random.sample(cells2.tolist(), num_to_choose)\n",
    "        \n",
    "#         if 0 not in beginning:\n",
    "#             beginning.insert(0, 0)\n",
    "#         if min_path not in ending:\n",
    "#             ending.append(min_path)\n",
    "#         indices = beginning + ending + cells3.tolist()\n",
    "\n",
    "        \n",
    "\n",
    "    # Plotting using multiple line plots\n",
    "    import matplotlib.patches as mpatches\n",
    "\n",
    "    #myColors = ((248/256,177/256,149/256),(246/256,114/256,128/256),(192/256,108/256,132/256),(108/256,91/256,123/256),(53/256,92/256,125/256))\n",
    "    \n",
    "    first_patch = mpatches.Patch(color=cmap(0/5), label=phases[0])\n",
    "    sec_patch = mpatches.Patch(color=cmap(1/5), label=phases[1])\n",
    "    third_patch = mpatches.Patch(color=cmap(2/5), label=phases[2])\n",
    "    fourth_patch = mpatches.Patch(color=cmap(3/5), label=phases[3])\n",
    "    fifth_patch = mpatches.Patch(color=cmap(4/5), label=phases[4])\n",
    "    plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "    \n",
    "#     first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "#     sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "#     third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "#     fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "#     fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "#     plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "    ax = plt.axes()\n",
    "    ax.set_facecolor(\"white\")\n",
    "\n",
    "#     cell = 0\n",
    "#     for i in range(len(interval_cell_progression)):\n",
    "#         for j in range(len(interval_cell_progression[i])):\n",
    "#             plt.plot([j, j+1], [cell, cell], color = myColors[interval_cell_progression[i][j]], linewidth=line_width)\n",
    "#         cell = cell + 1\n",
    "#     #plt.ylabel(\"Cells\")\n",
    "#     plt.xlabel(\"Trajectory\")\n",
    "#     plt.yticks([])\n",
    "#     plt.xlim(-1, max(path_len)+1)\n",
    "\n",
    "#     cell = 0\n",
    "#     for i in range(len(cell_progression)):\n",
    "#         for j in range(len(cell_progression[i])):\n",
    "#             plt.plot([j, j+1], [cell, cell], color = myColors[cell_progression[i][j]], linewidth=line_width)\n",
    "#         cell = cell + 1\n",
    "    #plt.ylabel(\"Cells\")\n",
    "    plt.xlabel(\"Trajectory\")\n",
    "    plt.yticks([])\n",
    "    #plt.xlim(-1, max(path_len)+1)\n",
    "\n",
    "#     cell_range = paths[0][paths[0].shape[0]-1,0] - paths[0][0,0]\n",
    "#     interval = cell_range/4\n",
    "#     cells3 = np.where((paths[0][:,0] > intermediate_coord[0]-interval)&(paths[0][:,0] < intermediate_coord[0]+interval))[0]\n",
    "    \n",
    "    #bounds = [0, np.min(cells3), np.max(cells3), paths[0].shape[0]-1]\n",
    "    #bounds = [0,20, 40, max(path_len)-1]\n",
    "    #bounds = [-10, 0, min_intermediate, 20]\n",
    "    #print(bounds)\n",
    "\n",
    "#     n_clusters = 3\n",
    "#     cmap = plt.get_cmap('Set3', n_clusters)\n",
    "#     norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "#     sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "#     #sm.set_array([])\n",
    "#     cbar = plt.colorbar(sm,\n",
    "#                  orientation=\"horizontal\")\n",
    "#     #tick_locs = (np.arange(3) + 0.5)*(3-1)/3\n",
    "#     tick_locs = [(0+bounds[1]/2, (bounds[1]+bounds[2]/2), (bounds[2]+bounds[3])/2)]\n",
    "#     cbar.set_ticks(tick_locs)\n",
    "#     cbar.set_ticklabels(['E','I','M'])\n",
    "\n",
    "#     fig, ax = plt.subplots(figsize=(6, 1))\n",
    "#     fig.subplots_adjust(bottom=0.6)\n",
    "    \n",
    "#     cmap = plt.get_cmap('Set3', 3)\n",
    "#     norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "#     cbar = fig.colorbar(\n",
    "#         mpl.cm.ScalarMappable(cmap=cmap, norm=norm),\n",
    "#         cax=ax,\n",
    "#         spacing='proportional',\n",
    "#         orientation='horizontal')\n",
    "    \n",
    "#     tick_locs = [(0+bounds[1]/2, (bounds[1]+bounds[2])/2, (bounds[2]+bounds[3])/2)]\n",
    "#     cbar.set_ticklabels(['Epithelial','Intermediate','Mesenchymal'])\n",
    "    \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "65b56d5c",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_cell_progression_lineplot3(cell_progression, paths, mid, line_width=10):\n",
    "    \n",
    "    phases = np.array(['M-G1','G1-S', 'S','G2-M','M'], dtype=object)\n",
    "    myColors = ((255/256,91/256,165/256),(219/256,73/256,172/256), (176/256,124/256,218/256), (107/256,167/256,214/256),(100/256,212/256,199/256))\n",
    "    \n",
    "    for i in range(paths.shape[0]):\n",
    "        coords = np.vstack((paths[i][:,0],np.repeat(0,paths[i].shape[0]))).T\n",
    "        index = getNNeighbors([mid, 0], coords, 1)[0][0]\n",
    "        #print(paths[i].shape[0], index)\n",
    "        x_val = -index\n",
    "        #print(len(cell_progression[i]))\n",
    "        #print(paths[i].shape[0])\n",
    "        for j in range(paths[i].shape[0]-1):\n",
    "            #print(cell_progression[i][j])\n",
    "            #print(x_val, cell_progression[i][j])\n",
    "            #print(paths[i].shape[0], j)\n",
    "            plt.plot([x_val, x_val + 1], [i,i], color=myColors[cell_progression[i][j]], linewidth=line_width)\n",
    "            x_val = x_val + 1\n",
    "    first_patch = mpatches.Patch(color=myColors[0], label=phases[0])\n",
    "    sec_patch = mpatches.Patch(color=myColors[1], label=phases[1])\n",
    "    third_patch = mpatches.Patch(color=myColors[2], label=phases[2])\n",
    "    fourth_patch = mpatches.Patch(color=myColors[3], label=phases[3])\n",
    "    fifth_patch = mpatches.Patch(color=myColors[4], label=phases[4])\n",
    "    plt.legend(handles=[first_patch, sec_patch, third_patch, fourth_patch, fifth_patch], loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "    plt.ylabel(\"Cells\")\n",
    "    plt.xlabel(\"Trajectory\")\n",
    "    plt.yticks([])\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fab98bf",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9da1bf24",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "6757bf3e",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_heatmap(cell_progression, paths, intermediate_coord):\n",
    "    lengths = []\n",
    "    phases = np.array(['M-G1','G1-S', 'S','G2-M','M'], dtype=object)\n",
    "    for i in range(cell_progression.shape[0]):\n",
    "        lengths.append(len(cell_progression[i]))\n",
    "    min_length = min(lengths)    \n",
    "    \n",
    "    #print(lengths)\n",
    "    #middle_coord = []\n",
    "    \n",
    "    cell_prog = []\n",
    "    cell_progression2 = cell_progression\n",
    "    for i in range(cell_progression2.shape[0]):\n",
    "        index = getNNeighbors(intermediate_coord, paths[i], 1)[0][0]\n",
    "        #middle_coord.append(index)\n",
    "        \n",
    "        #print(len(cell_progression2[i]) - min_length)\n",
    "        \n",
    "        temp = cell_progression2[i]\n",
    "        for j in range(len(cell_progression2[i]) - min_length):\n",
    "            num = random.randint(0, len(temp)-1)\n",
    "            temp.pop(num)\n",
    "        cell_prog.append(temp)\n",
    "        #print(temp, \"\\n\")\n",
    "        \n",
    "    #print(middle_coord)\n",
    "    #print(np.average(np.array(middle_coord)))\n",
    "    \n",
    "    cmap = sns.color_palette(\"rocket\", phases.shape[0])\n",
    "    ax = sns.heatmap(cell_prog, cmap=cmap)\n",
    "    colorbar = ax.collections[0].colorbar\n",
    "    r = colorbar.vmax - colorbar.vmin\n",
    "    colorbar.set_ticks([colorbar.vmin + 0.5 * r / (phases.shape[0]) + r * i / (phases.shape[0]) for i in range(phases.shape[0])])\n",
    "    colorbar.set_ticklabels(phases)\n",
    "    ax.set_ylabel('Cells')\n",
    "    ax.set_xlabel('Trajectory')\n",
    "    ax.set(yticklabels=[])\n",
    "    ax.set(xticklabels=[])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "60e62ff4",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def get_ge(adata, path, gene_name):\n",
    "    ge = []\n",
    "    gene_index = np.where(adata.var['var_names'] == gene_name)[0][0]\n",
    "    #print(path.shape[0])\n",
    "    for i in range(path.shape[0]):\n",
    "        indices = getNNeighbors(path[i], adata.obsm['X_umap'], 20)\n",
    "        temp = adata.X[indices, gene_index].toarray()\n",
    "        ge.append(np.average(temp))\n",
    "    return ge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6407c398",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def get_ge_paths(adata, paths, gene_name):\n",
    "    ge = []\n",
    "    for i in range(paths.shape[0]):\n",
    "        ge.append(get_ge(adata, paths[i], gene_name))\n",
    "    return ge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "f9e4eade",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_heatmap_ge(ge, paths, gene):\n",
    "    lengths = []\n",
    "    for i in range(paths.shape[0]):\n",
    "        lengths.append(len(paths[i]))\n",
    "    min_length = min(lengths)    \n",
    "\n",
    "    final_ge = []\n",
    "    for i in range(paths.shape[0]):\n",
    "        temp = ge[i]\n",
    "        for j in range(len(ge[i]) - min_length):\n",
    "            num = random.randint(0, len(temp)-1)\n",
    "            temp.pop(num)\n",
    "        final_ge.append(temp)\n",
    "        \n",
    "    ax = sns.heatmap(final_ge, cmap=sns.color_palette(\"rocket\", as_cmap=True))\n",
    "    ax.set_ylabel('Cells')\n",
    "    ax.set_xlabel('Trajectory')\n",
    "    ax.set(yticklabels=[])\n",
    "    #ax.set(xticklabels=[])\n",
    "    ax.set_title(gene + \" trajectory expression\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27d7747d",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bb79b20",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3e87822",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ec4c0ea2",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_traj_states(paths, adata):\n",
    "    #Plotting trajectories into 3 states (epithelial, intermediate, mesenchymal)\n",
    "    from matplotlib import cm\n",
    "    n_clusters = 3\n",
    "    #cmap = plt.get_cmap('Set3', n_clusters)\n",
    "    cmap = colors.ListedColormap([(0, 85/256, 129/256), (114/256,205/256,244/256), (255/256, 210/256, 0)])\n",
    "    states = ['Epithelial', 'Intermediate','Mesenchymal']\n",
    "\n",
    "    for j in range(paths.shape[0]):\n",
    "        final_coords = []\n",
    "\n",
    "        x_coords = np.linspace(paths[j][0,0], paths[j][paths[j].shape[0]-1,0], 30)\n",
    "        coords = np.vstack((paths[j][:,0], np.repeat(0, paths[j].shape[0]))).T\n",
    "        interval_coords = np.vstack((x_coords, np.repeat(0, x_coords.shape[0]))).T\n",
    "\n",
    "        plt.scatter(adata.obsm[\"X_umap\"][:,0], adata.obsm[\"X_umap\"][:,1], color = \"lightgrey\")\n",
    "\n",
    "        for i in range(interval_coords.shape[0]): \n",
    "            index = getNNeighbors(interval_coords[i], coords, 1)\n",
    "\n",
    "            final_coords.append(paths[j][index,:][0][0])\n",
    "        final_coords = np.array(final_coords)\n",
    "        interval = np.linspace(final_coords[0,0], final_coords[len(final_coords)-1,0], 4)\n",
    "\n",
    "        ax = plt.axes()\n",
    "        ax.set_facecolor(\"white\")\n",
    "\n",
    "        for i in range(1,len(interval)):\n",
    "            indices = np.where((final_coords[:,0] < interval[i]) & (final_coords[:,0] >= interval[i-1]))[0]\n",
    "            plt.scatter(final_coords[indices,0], final_coords[indices,1], color=cmap(i-1), label = states[i-1])\n",
    "        plt.scatter(final_coords[0,0], final_coords[0,1], marker = \"*\", color = \"red\", label=\"start\")\n",
    "        plt.scatter(final_coords[final_coords.shape[0]-1,0], final_coords[final_coords.shape[0]-1,1], marker = \"*\", color = \"black\", label = \"end\")\n",
    "        plt.legend(loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "        plt.axis('off')\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d17eee3c",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Splits the trajectories into even thirds and plots the corresponding state colors \n",
    "def plot_traj_states2(paths, adata, intermediate_coord):\n",
    "    #Plotting trajectories into 3 states (epithelial, intermediate, mesenchymal)\n",
    "    from matplotlib import cm\n",
    "    n_clusters = 3\n",
    "    #cmap = plt.get_cmap('Set3', n_clusters)\n",
    "    cmap = colors.ListedColormap([(0, 85/256, 129/256), (114/256,205/256,244/256), (255/256, 210/256, 0)])\n",
    "    states = ['Epithelial', 'Intermediate','Mesenchymal']\n",
    "\n",
    "    for i in range(paths.shape[0]):\n",
    "            \n",
    "        #state_coords = [paths[i][0][0], intermediate_coord, paths[i][paths[i].shape[0]-1]]    \n",
    "\n",
    "        cells = int(paths[i].shape[0]/3)\n",
    "        indices = getNNeighbors(intermediate_coord, paths[i], k=cells)[0]\n",
    "\n",
    "        min_indices = np.min(indices)\n",
    "        max_indices = np.max(indices)\n",
    "\n",
    "        print(min_indices, max_indices, paths[i].shape[0])\n",
    "        \n",
    "        plt.scatter(adata.obsm[\"X_umap\"][:,0], adata.obsm[\"X_umap\"][:,1], color = \"lightgrey\")\n",
    "\n",
    "        plt.scatter(paths[i][0:min_indices,0], paths[i][0:min_indices,1], color = cmap(0))\n",
    "        plt.scatter(paths[i][min_indices:max_indices,0], paths[i][min_indices:max_indices,1], color = cmap(1))\n",
    "        plt.scatter(paths[i][max_indices:paths[i].shape[0]-1,0], paths[i][max_indices:paths[i].shape[0]-1,1], color = cmap(2))\n",
    "\n",
    "        plt.scatter(intermediate_coord[0], intermediate_coord[1], marker = '*', color = 'blue', label = 'intermediate')\n",
    "        plt.scatter(paths[i][0,0], paths[i][0,1], marker = \"*\", color = \"red\", label=\"start\")\n",
    "        plt.scatter(paths[i][paths[i].shape[0]-1,0], paths[i][paths[i].shape[0]-1,1], marker = \"*\", color = \"black\", label = \"end\")\n",
    "        plt.legend(loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "        plt.axis('off')\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "fe0d6bb7",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Splits the trajectories into thirds for each state\n",
    "# Arbitrarily assigned a range for the intermediate state \n",
    "# All the other states considered the epithelial or mesenchymal state\n",
    "\n",
    "def plot_traj_states3(paths, adata, intermediate_coord):\n",
    "    #Plotting trajectories into 3 states (epithelial, intermediate, mesenchymal)\n",
    "    from matplotlib import cm\n",
    "    n_clusters = 3\n",
    "    #cmap = plt.get_cmap('Set3', n_clusters)\n",
    "    cmap = colors.ListedColormap([(0, 85/256, 129/256), (114/256,205/256,244/256), (255/256, 210/256, 0)])\n",
    "    states = ['Epithelial', 'Intermediate','Mesenchymal']\n",
    "\n",
    "    for i in range(paths.shape[0]):\n",
    "        \n",
    "        cell_range = paths[i][paths[i].shape[0]-1,0] - paths[i][0,0]\n",
    "        #print(cell_range)\n",
    "        print(paths[i].shape[0])\n",
    "        interval = cell_range/3\n",
    "        #print(interval)\n",
    "        #print(intermediate_coord[0]-interval, intermediate_coord[0]+interval)\n",
    "        \n",
    "        cells = np.where(paths[i][:,0] < intermediate_coord[0]-interval)[0]\n",
    "        cells2 = np.where(paths[i][:,0] > intermediate_coord[0]+interval)[0]\n",
    "        cells3 = np.where((paths[i][:,0] > intermediate_coord[0]-interval)&(paths[i][:,0] < intermediate_coord[0]+interval))[0]\n",
    "        outside_cells = np.concatenate((cells, cells2))\n",
    "        \n",
    "        plt.scatter(adata.obsm[\"X_umap\"][:,0], adata.obsm[\"X_umap\"][:,1], color = \"lightgrey\")\n",
    "        \n",
    "        plt.scatter(paths[i][cells,0], paths[i][cells,1], color = cmap(0))\n",
    "        plt.scatter(paths[i][cells2,0], paths[i][cells2,1], color = cmap(1))\n",
    "        plt.scatter(paths[i][cells3,0], paths[i][cells3,1], color = cmap(2))\n",
    "        \n",
    "        for j in range(0,paths[i].shape[0],10):\n",
    "            plt.scatter(paths[i][j,0], paths[i][j,1], color = 'black')\n",
    "        \n",
    "        plt.scatter(intermediate_coord[0], intermediate_coord[1], marker = '*', color = 'blue', label = 'intermediate')\n",
    "        plt.scatter(paths[i][0,0], paths[i][0,1], marker = \"*\", color = \"red\", label=\"start\")\n",
    "        plt.scatter(paths[i][paths[i].shape[0]-1,0], paths[i][paths[i].shape[0]-1,1], marker = \"*\", color = \"black\", label = \"end\")\n",
    "        plt.legend(loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "        plt.axis('off')\n",
    "        plt.show()\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d00cd2",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62eded16",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "448dea2b",
   "metadata": {
    "collapsed": true,
    "hidden": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def plot_changing_gene_exp(gene_names, gene_regulation, groups, filename, threshold = 2, plot_all = True):\n",
    "    changing_genes = []\n",
    "    index = 1\n",
    "\n",
    "    timepoints = np.repeat([range(len(gene_regulation[0]))], len(gene_names), axis=0).tolist()\n",
    "    \n",
    "    for i in range(len(gene_regulation)):\n",
    "        if plot_all == False:\n",
    "            if np.unique(gene_regulation[i]).shape[0] > 1:\n",
    "                if (np.max(gene_regulation[i]) - np.min(gene_regulation[i])) >= threshold:\n",
    "                    plt.plot(timepoints[i], gene_regulation[i], marker = 'o')\n",
    "                    plt.title(gene_names[i])\n",
    "                    plt.ylim(-0.5, groups+0.5-1)\n",
    "                    index = index + 1\n",
    "                    changing_genes.append(gene_names[i])\n",
    "                    plt.savefig(\"Plots/EMT_trajectory/\" + filename + \"/\"+gene_names[i]+\"_progression\", dpi=300)\n",
    "                    plt.show()\n",
    "        else:\n",
    "            plt.plot(timepoints[i], gene_regulation[i], marker = 'o')\n",
    "            plt.title(gene_names[i])\n",
    "            plt.ylim(-0.5, groups+0.5-1)\n",
    "            index = index + 1\n",
    "            changing_genes.append(gene_names[i])\n",
    "            plt.savefig(\"Plots/EMT_trajectory/\" + filename + \"/\"+gene_names[i]+\"_progression\", dpi=300)\n",
    "            plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "13b57950",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def get_cell_ge(path_sol, gene_index, adata):\n",
    "    gene_expression = []\n",
    "    #indices = getNNeighbors(path_sol[i], adata.obsm['X_umap'], k=100)[0]\n",
    "    \n",
    "    #for i in range(len(gene_names)):\n",
    "        #temp = []\n",
    "    #print(indexes)\n",
    "    #print(gene_names)\n",
    "    for j in range(0,len(path_sol),20):\n",
    "        indices = getNNeighbors(path_sol[j], adata.obsm['X_umap'], k=100)[0]\n",
    "        temp = []\n",
    "        for k in range(len(indices)):\n",
    "            temp.append(adata.X[indices[k], gene_index])\n",
    "#         temp_mean = []\n",
    "#         for i in range(0, len(temp), 10):\n",
    "#             if i+10 > len(temp):\n",
    "#                 temp_mean.append(mean(temp[i:]))\n",
    "#             else:\n",
    "#                 temp_mean.append(mean(temp[i:i+10]))\n",
    "        gene_expression.append(temp)\n",
    "    return gene_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c61be502",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_ge(gene_name, gene_expression):\n",
    "    x_axis = np.arange(0, len(gene_expression)*len(gene_expression[0]), 1)\n",
    "    for i in range(len(gene_expression)):\n",
    "        #print(len(x_axis[i*len(gene_expression[i]):(i+1)*len(gene_expression[i])]))\n",
    "        #print(len(gene_expression[i]))\n",
    "        plt.bar(x_axis[i*len(gene_expression[i]):(i+1)*len(gene_expression[i])], gene_expression[i], color=(0.1, 0.1, 0.1, 0.1),  edgecolor='blue')\n",
    "        plt.title(gene_name + \" gene expression\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d8256bd9",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plotting_ge_trajectory(path_sol, gene_index, gene_name, adata):\n",
    "    gene_expression = []\n",
    "    \n",
    "    for j in range(0,len(path_sol)):\n",
    "        indices = getNNeighbors(path_sol[j], adata.obsm['X_umap'], k=100)[0]\n",
    "        temp = 0\n",
    "        for k in range(len(indices)):\n",
    "            temp = temp + adata.X[indices[k], gene_index]\n",
    "        gene_expression.append(temp/100)\n",
    "    x_axis = np.arange(0, len(gene_expression), 1)\n",
    "    # Plotting bar plot\n",
    "    plt.bar(x_axis, gene_expression, color=(0.1, 0.1, 0.1, 0.1),  edgecolor='blue')\n",
    "    \n",
    "    plt.title(gene_name)\n",
    "    plt.ylim(0,np.max(adata.X.toarray()))\n",
    "    plt.show()\n",
    "\n",
    "    #return gene_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6aa4fed0",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plotting_ge_trajectory_scatter(path_sol, gene_index, gene_name, adata, cells):\n",
    "    gene_expression = []\n",
    "    \n",
    "    for j in range(0,len(path_sol),3):\n",
    "        indices = getNNeighbors(path_sol[j], adata.obsm['X_umap'], k=cells)[0]\n",
    "        #print(np.repeat(j, cells).shape)\n",
    "        #print(adata.X.toarray()[indices, gene_index].shape)\n",
    "        plt.scatter(np.repeat(j, cells), adata.X.toarray()[indices, gene_index])\n",
    "        \n",
    "    #x_axis = np.arange(0, len(gene_expression), 1)\n",
    "    # Plotting bar plot\n",
    "    #plt.bar(x_axis, gene_expression, color=(0.1, 0.1, 0.1, 0.1),  edgecolor='blue')\n",
    "    \n",
    "    plt.title(gene_name)\n",
    "    #plt.ylim(0,np.max(adata.X.toarray()))\n",
    "    plt.show()\n",
    "\n",
    "    #return gene_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "dd3f9a17",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_dendrogram(model, **kwargs):\n",
    "    # Create linkage matrix and then plot the dendrogram\n",
    "\n",
    "    # create the counts of samples under each node\n",
    "    counts = np.zeros(model.children_.shape[0])\n",
    "    n_samples = len(model.labels_)\n",
    "    for i, merge in enumerate(model.children_):\n",
    "        current_count = 0\n",
    "        for child_idx in merge:\n",
    "            if child_idx < n_samples:\n",
    "                current_count += 1  # leaf node\n",
    "            else:\n",
    "                current_count += counts[child_idx - n_samples]\n",
    "        counts[i] = current_count\n",
    "\n",
    "    linkage_matrix = np.column_stack([model.children_, model.distances_,\n",
    "                                      counts]).astype(float)\n",
    "\n",
    "    # Plot the corresponding dendrogram\n",
    "    dendrogram(linkage_matrix, **kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d2c79f8",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6e063a3",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "35a31b18",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def louvain_clustering(adata, res):\n",
    "    dose = np.unique(adata.obs['dosage'])[0]\n",
    "    print(dose, \":\")\n",
    "    sc.tl.louvain(adata, resolution=res, key_added='louvain_r'+str(res), random_state=10)\n",
    "    print(adata.obs['louvain_r'+str(res)].value_counts())\n",
    "    print(\"\\n\")\n",
    "    sc.pl.umap(adata, color=['louvain_r'+str(res)], palette=sc.pl.palettes.vega_20)\n",
    "    sc.tl.rank_genes_groups(adata, groupby='louvain_r'+str(res), key_added='rank_genes_r' + str(res), groups=[0,1])\n",
    "    sc.pl.rank_genes_groups(adata, key='rank_genes_r' + str(res), groups=['0'], fontsize=12)\n",
    "    sc.pl.rank_genes_groups(adata, key='rank_genes_r' + str(res), groups=['1'], fontsize=12)\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "62704d05",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plotNormalizedCounts(data_df, gene):\n",
    "    rc_unique = data_df['RC nneighbor'].unique()[np.argsort(data_df['RC nneighbor'].unique())]\n",
    "    rc_mean_exp = []\n",
    "    for i in rc_unique:\n",
    "        temp = data_df[data_df['RC nneighbor']==i]\n",
    "        mean_num = np.mean(temp[gene])\n",
    "        rc_mean_exp.append(mean_num)\n",
    "    rc_mean_exp = pd.Series(rc_mean_exp).fillna(0).tolist()\n",
    "    #z = np.polyfit(rc_unique, rc_mean_exp, 8)\n",
    "    #p = np.poly1d(z)\n",
    "    #xp = np.linspace(1, 109, 104)\n",
    "\n",
    "    fig,ax=plt.subplots(figsize=(7.5,5), dpi=300)\n",
    "    plt.scatter(data_df['RC nneighbor'], data_df[gene], c=data_df['CC'].map(colors))\n",
    "    plt.hlines(0, xmin=-1, xmax=115, linestyles='dashed')\n",
    "    #plt.plot(rc_unique, p(xp), color='black')\n",
    "    plt.xlabel('Reaction coordinate',fontsize=12,fontweight='bold')\n",
    "    plt.ylabel('Normalized counts', fontsize=12,fontweight='bold')\n",
    "    plt.xticks(fontsize=10,fontweight='bold')\n",
    "    plt.yticks(fontsize=10,fontweight='bold')\n",
    "    plt.margins(0.05)\n",
    "    plt.title(gene+' expression along reaction coordinate', fontsize=14,fontweight='bold')\n",
    "    plt.savefig(\"result/RC_exp/normalized/\"+gene+\".png\", transparent=True, dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18c2588a",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plotOrigCounts(data_df, gene):\n",
    "    rc_unique = data_df['RC nneighbor'].unique()[np.argsort(data_df['RC nneighbor'].unique())]\n",
    "    rc_mean_exp = []\n",
    "    for i in rc_unique:\n",
    "        temp = data_df[data_df['RC nneighbor']==i]\n",
    "        mean_num = np.mean(temp[gene])\n",
    "        rc_mean_exp.append(mean_num)\n",
    "    rc_mean_exp = pd.Series(rc_mean_exp).fillna(0).tolist()\n",
    "    z = np.polyfit(rc_unique, rc_mean_exp, 8)\n",
    "    p = np.poly1d(z)\n",
    "    xp = np.linspace(1, 109, 104)\n",
    "\n",
    "    fig,ax=plt.subplots(figsize=(7.5,5), dpi=300)\n",
    "    plt.scatter(data_df['RC nneighbor'], data_df[gene], c=data_df['CC'].map(colors))\n",
    "    #plt.hlines(0, xmin=-1, xmax=115, linestyles='dashed')\n",
    "    plt.plot(rc_unique, p(xp), color='black')\n",
    "    plt.xlabel('Reaction coordinate',fontsize=12,fontweight='bold')\n",
    "    plt.ylabel('Original counts', fontsize=12,fontweight='bold')\n",
    "    plt.xticks(fontsize=10,fontweight='bold')\n",
    "    plt.yticks(fontsize=10,fontweight='bold')\n",
    "    plt.margins(0.05)\n",
    "    plt.title(gene+' expression along reaction coordinate', fontsize=14,fontweight='bold')\n",
    "    plt.savefig(\"result/RC_exp/\"+gene+\".png\", transparent=True, dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df56c257",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69465171",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "52508c9b",
   "metadata": {
    "heading_collapsed": true,
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## DTW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "4a25801b",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Performs DTW using the first path in the paths array as the reference\n",
    "# Returns the DTW distance and the warp_paths in a list\n",
    "def align_DTW(paths, ref_path):\n",
    "    warp_paths = []\n",
    "    distances = []\n",
    "\n",
    "    for i in range(0, paths.shape[0]):\n",
    "        x = ref_path[:,1]\n",
    "        y = paths[i][:,1]\n",
    "        dtw_distance, warp_path = fastdtw(x, y, dist=euclidean)\n",
    "        warp_paths.append(warp_path)\n",
    "        distances.append(dtw_distance)\n",
    "    return distances, warp_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "84f366fc",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Plots the new trajectories \n",
    "def plot_warped_paths(paths, warp_paths, ref_path, adata):\n",
    "    toReturn = []\n",
    "    for i in range(len(warp_paths)):\n",
    "        new_path = []\n",
    "        index = 0\n",
    "        while index < len(warp_paths[i])-1:\n",
    "            x1, y1 = warp_paths[i][index]\n",
    "            #total = paths[i+1][y1,1]\n",
    "            total = paths[i][y1,1]\n",
    "            num = 1\n",
    "            cont_bol = True\n",
    "            index += 1\n",
    "            \n",
    "            while cont_bol & (index <= len(warp_paths[i])-1):\n",
    "                x2, y2 = warp_paths[i][index]\n",
    "                if x1 == x2:\n",
    "                    #total += paths[i+1][y2,1]\n",
    "                    total += paths[i][y2,1]\n",
    "                    index = index + 1\n",
    "                    num+=1\n",
    "                else:\n",
    "                    cont_bol = False\n",
    "            new_path.append(total/num)\n",
    "            \n",
    "        #if len(new_path)!= paths[0].shape[0]:\n",
    "        if len(new_path)!= ref_path.shape[0]:\n",
    "            x, y = warp_paths[i][len(warp_paths[i])-1]\n",
    "            #print(new_path.shape)\n",
    "            #new_path.append(paths[i+1][y,1])\n",
    "            new_path.append(paths[i][y,1])\n",
    "        print(i)\n",
    "        plt.scatter(adata.obsm[\"X_umap\"][:,0], adata.obsm[\"X_umap\"][:,1], color = \"lightgrey\")\n",
    "        #plt.scatter(paths[0][:,0], new_path)\n",
    "        plt.scatter(ref_path[:,0], new_path)\n",
    "        #plt.scatter(paths[i+1][:,0], paths[i+1][:,1], c=paths[i+1][:,0])\n",
    "        plt.scatter(paths[i][:,0], paths[i][:,1], c=paths[i][:,0])\n",
    "        plt.show()\n",
    "        \n",
    "        toReturn.append(np.vstack((ref_path[:,0], new_path)).T)\n",
    "        \n",
    "    return np.array(toReturn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60eb6df0",
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mRunning cells with 'Python 3.7.8 64-bit' requires ipykernel package.\n",
      "Run the following command to install 'ipykernel' into the Python environment. \n",
      "Command: '\"C:/Program Files (x86)/Microsoft Visual Studio/Shared/Python37_64/python.exe\" -m pip install ipykernel -U --user --force-reinstall'"
     ]
    }
   ],
   "source": [
    "# Plots the new trajectories \n",
    "def plot_warped_paths2(paths, warp_paths, ref_path, adata):\n",
    "    toReturn = []\n",
    "    for i in range(len(warp_paths)):\n",
    "        new_path = []\n",
    "        index = 0\n",
    "        while index < len(warp_paths[i])-1:\n",
    "            x1, y1 = warp_paths[i][index]\n",
    "            #total = paths[i+1][y1,1]\n",
    "            total = paths[i][y1,1]\n",
    "            num = 1\n",
    "            cont_bol = True\n",
    "            index += 1\n",
    "            \n",
    "            while cont_bol & (index <= len(warp_paths[i])-1):\n",
    "                x2, y2 = warp_paths[i][index]\n",
    "                if x1 == x2:\n",
    "                    #total += paths[i+1][y2,1]\n",
    "                    total += paths[i][y2,1]\n",
    "                    index = index + 1\n",
    "                    num+=1\n",
    "                else:\n",
    "                    cont_bol = False\n",
    "            new_path.append(total/num)\n",
    "            \n",
    "        #if len(new_path)!= paths[0].shape[0]:\n",
    "        if len(new_path)!= ref_path.shape[0]:\n",
    "            x, y = warp_paths[i][len(warp_paths[i])-1]\n",
    "            #print(new_path.shape)\n",
    "            #new_path.append(paths[i+1][y,1])\n",
    "            new_path.append(paths[i][y,1])\n",
    "        # print(i)\n",
    "        # plt.scatter(adata.obsm[\"X_umap\"][:,0], adata.obsm[\"X_umap\"][:,1], color = \"lightgrey\")\n",
    "        # #plt.scatter(paths[0][:,0], new_path)\n",
    "        # plt.scatter(ref_path[:,0], new_path)\n",
    "        # #plt.scatter(paths[i+1][:,0], paths[i+1][:,1], c=paths[i+1][:,0])\n",
    "        # plt.scatter(paths[i][:,0], paths[i][:,1], c=paths[i][:,0])\n",
    "        # plt.show()\n",
    "        \n",
    "        toReturn.append(np.vstack((ref_path[:,0], new_path)).T)\n",
    "        \n",
    "    return np.array(toReturn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "7b87a67e",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Calculates pairwise distances between list of trajectories \n",
    "def pairwise_DTW(paths):\n",
    "    distances = []\n",
    "    warp_paths = []\n",
    "    for i in range(paths.shape[0]):\n",
    "        dist_temp = []\n",
    "        warp_temp = []\n",
    "        for j in range(paths.shape[0]):\n",
    "            x = paths[i][:,1]\n",
    "            y = paths[j][:,1]\n",
    "            dtw_distance, warp_path = fastdtw(x, y, dist=euclidean)\n",
    "            dist_temp.append(dtw_distance)\n",
    "            warp_temp.append(warp_path)\n",
    "        distances.append(dist_temp)\n",
    "        warp_paths.append(warp_temp)\n",
    "    return distances, warp_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dde93c24",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Returns the index of the trajectories that are the closest in distance\n",
    "def min_dist(distances):\n",
    "    min_list = []\n",
    "    for i in range(len(distances)):\n",
    "        temp = np.array(distances[i])\n",
    "        order = np.argsort(temp)\n",
    "        min_list.append(temp[order][1])\n",
    "    traj1_index = min_list.index(min(min_list))\n",
    "    #traj2_index = distances[traj1_index].index(min(min_list))\n",
    "    traj2_index = np.where(np.array(distances[traj1_index]==min(min_list)))[0]\n",
    "    if len(traj2_index) > 1:\n",
    "        traj2_index = traj2_index[1]\n",
    "    else:\n",
    "        traj2_index = traj2_index[0]\n",
    "    return traj1_index, traj2_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bfe2a40",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Calculates the average of 2 trajectories \n",
    "def calc_avg_traj(path1, path2):\n",
    "    toReturn = []\n",
    "    for i in range(path2.shape[0]):\n",
    "        toReturn.append((path1[i,1] + path2[i,1])/2)\n",
    "    toReturn = np.vstack((path1[:,0], np.array(toReturn)))\n",
    "    return toReturn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68500669",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Obtains the new warped path for a single trajectory\n",
    "def get_warped_path(paths, warp_path):\n",
    "    new_path = []\n",
    "    index = 0\n",
    "    while index < len(warp_path)-1:\n",
    "        x1, y1 = warp_path[index]\n",
    "        total = paths[1][y1,1] # get traj y value\n",
    "        num = 1\n",
    "        cont_bol = True\n",
    "        index += 1\n",
    "\n",
    "        while cont_bol & (index <= len(warp_path)-1):\n",
    "            x2, y2 = warp_path[index]\n",
    "            # when multiple points are mapped to the same point\n",
    "            if x1 == x2:\n",
    "                total += paths[1][y2,1]\n",
    "                index = index + 1\n",
    "                num+=1\n",
    "            else:\n",
    "                cont_bol = False\n",
    "        new_path.append(total/num)\n",
    "\n",
    "    if len(new_path)!= paths[0].shape[0]:\n",
    "        x, y = warp_path[len(warp_path)-1]\n",
    "        new_path.append(paths[1][y,1])\n",
    "        \n",
    "    return np.array((paths[0][:,0], new_path)).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2175765a",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Obtains the new warped path for a single trajectory\n",
    "# paths\n",
    "# warp_path\n",
    "# expr : num time points x num of genes\n",
    "def get_warped_expr(paths, warp_path, expr):\n",
    "    new_path = []\n",
    "    new_expr = []\n",
    "    index = 0\n",
    "    while index < len(warp_path)-1:\n",
    "        x1, y1 = warp_path[index]\n",
    "        #print('y1:', y1)\n",
    "        total = paths[1][y1,1] # get traj y value\n",
    "        curr_expr = expr[y1,:]\n",
    "        num = 1\n",
    "        cont_bol = True\n",
    "        index += 1\n",
    "\n",
    "        while cont_bol & (index <= len(warp_path)-1):\n",
    "            x2, y2 = warp_path[index]\n",
    "            #print('\\ty2:', y2)\n",
    "            # when multiple points are mapped to the same point\n",
    "            if x1 == x2:\n",
    "                total += paths[1][y2,1]\n",
    "                curr_expr += expr[y2, :]\n",
    "                index = index + 1\n",
    "                num+=1\n",
    "            else:\n",
    "                cont_bol = False\n",
    "        new_path.append(total/num)\n",
    "        new_expr.append(curr_expr/num)\n",
    "\n",
    "    if len(new_path)!= paths[0].shape[0]:\n",
    "        x, y = warp_path[len(warp_path)-1]\n",
    "        new_path.append(paths[1][y,1])\n",
    "        new_expr.append(expr[y,:])\n",
    "        \n",
    "    return np.array((paths[0][:,0], new_path)).T, np.array(new_expr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "b99c9944",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "#Works for trajectory numbeers with multiples of 4 or 8\n",
    "def get_average_traj(paths):\n",
    "    #Find trajectories with the smallet distances between them\n",
    "    n = paths.shape[0]\n",
    "    final_avg = None\n",
    "\n",
    "    for i in range(int(np.log2(n)-1)):\n",
    "        # Pairs every trajectory using minimum distance\n",
    "        new_paths = []\n",
    "        n = paths.shape[0]\n",
    "        for j in range(int(n/2)):\n",
    "            # Getting closest pair of trajectories\n",
    "            dist_pairwise, warped_pairwise = pairwise_DTW(paths)\n",
    "            \n",
    "            traj1, traj2 = min_dist(dist_pairwise)\n",
    "            \n",
    "            # Gets new warped path to matched indices \n",
    "            _, warp_path = fastdtw(paths[traj1], paths[traj2], dist=euclidean)\n",
    "            new_path = get_warped_path([paths[traj1], paths[traj2]], warp_path)\n",
    "            \n",
    "            # Calculates average of the two paths\n",
    "            average_path = calc_avg_traj(paths[traj1], new_path)\n",
    "            new_paths.append(average_path.T)\n",
    "\n",
    "            if traj1 > traj2:\n",
    "                #print(\"traj1:\", traj1)\n",
    "                #print(\"traj2:\", traj2)\n",
    "                del dist_pairwise[traj1]\n",
    "                del dist_pairwise[traj2]\n",
    "                paths = np.delete(paths, traj1, 0)\n",
    "                paths = np.delete(paths, traj2, 0)\n",
    "            else:\n",
    "                #print(\"traj1:\", traj1)\n",
    "                #print(\"traj2:\", traj2)\n",
    "                del dist_pairwise[traj2]\n",
    "                del dist_pairwise[traj1]\n",
    "                paths = np.delete(paths, traj2, 0)\n",
    "                paths = np.delete(paths, traj1, 0)\n",
    "\n",
    "        paths = np.array(new_paths)\n",
    "\n",
    "        if len(new_paths) == 2:\n",
    "            _, warp_path = fastdtw(new_paths[0], new_paths[1], dist=euclidean)\n",
    "            new_path = get_warped_path([new_paths[0], new_paths[1]], warp_path)\n",
    "            final_avg = calc_avg_traj(new_paths[0], new_path)\n",
    "\n",
    "    return final_avg.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b2247f55",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Obtains the cell progression of the warped path\n",
    "# path[0] = avg path\n",
    "# path[1] = path being warped\n",
    "def get_cell_prog_warped_path(cell_prog, paths, warp_path):\n",
    "    phases = np.array(['M-G1','G1-S', 'S','G2-M','M'], dtype=object)\n",
    "    cell_prog_path = []\n",
    "    index = 0\n",
    "    while index < len(warp_path)-1:\n",
    "        x1, y1 = warp_path[index]\n",
    "        tracker = np.repeat(0, phases.shape[0])\n",
    "        tracker[cell_prog[y1]] = tracker[cell_prog[y1]] + 1\n",
    "        num = 1\n",
    "        cont_bol = True\n",
    "        index += 1\n",
    "\n",
    "        while cont_bol & (index <= len(warp_path)-1):\n",
    "            x2, y2 = warp_path[index]\n",
    "            if x1 == x2:\n",
    "                tracker[cell_prog[y2]] = tracker[cell_prog[y2]] + 1\n",
    "                index = index + 1\n",
    "                num+=1\n",
    "            else:\n",
    "                cont_bol = False\n",
    "        max_phase = np.where(tracker == np.max(tracker))[0][0]\n",
    "        cell_prog_path.append(max_phase)\n",
    "\n",
    "    if len(cell_prog_path)!= paths[0].shape[0]:\n",
    "        x, y = warp_path[len(warp_path)-1]\n",
    "        cell_prog_path.append(cell_prog[y])\n",
    "        \n",
    "    return np.array(cell_prog_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ae39125",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_cell_prog_DTW(cell_progressions, paths, avg_path, ):\n",
    "    for i in range(paths.shape[0]):\n",
    "        _, warped_path = fastdtw(avg_path, paths[i], dist=euclidean)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f49b1daa",
   "metadata": {
    "heading_collapsed": true,
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Jacobian"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cd2058a3",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Plots the jacobian of each cell in umap space\n",
    "def jacobian_cell_cell(adata, reg, eff):\n",
    "    dyn.pl.jacobian(adata, regulators=[reg], effectors=[eff], basis='umap')\n",
    "    dyn.pl.jacobian(adata, regulators=[eff], effectors=[reg], basis='umap')\n",
    "    dyn.pl.jacobian(adata, regulators=[reg], effectors=[reg], basis='umap')\n",
    "    dyn.pl.jacobian(adata, regulators=[eff], effectors=[eff], basis='umap')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c8027fd4",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Plots the jacobian of each cell as a function of gene concentration\n",
    "def jacobian_ge_level(adata, reg, eff):\n",
    "    dyn.pl.jacobian(adata, regulators=[reg], effectors=[eff], x=reg, y=eff, layer='M_s', basis='umap')\n",
    "    dyn.pl.jacobian(adata, regulators=[eff], effectors=[reg], x=eff, y=reg, layer='M_s', basis='umap')\n",
    "    #dyn.pl.jacobian(adata, regulators=[reg], effectors=[reg], x=reg, y=eff, layer='M_s', basis='umap')\n",
    "    #dyn.pl.jacobian(adata, regulators=[eff], effectors=[eff], x=eff, y=eff, layer='M_s', basis='umap')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "893e9c2b",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "#Looks at the top interactions\n",
    "def plot_jacobian_TF_reg(adata, reg, top_genes, dose, genes=None):\n",
    "    if genes == None:\n",
    "        dyn.pp.top_pca_genes(adata, n_top_genes=100)\n",
    "        top_pca_genes = adata.var.index[adata.var.top_pca_genes];\n",
    "        genes = top_pca_genes.tolist()\n",
    "        genes.append('SNAI1')\n",
    "        genes.append('SNAI2')\n",
    "        genes.append('TWIST1')\n",
    "        genes.append('TWIST2')\n",
    "        genes.append('ZEB1')\n",
    "        genes.append('ZEB2')\n",
    "    else:\n",
    "        top_genes = len(genes)\n",
    "    \n",
    "    dyn.vf.jacobian(adata, regulators=[reg] , effectors=genes)\n",
    "    dyn.vf.rank_jacobian_genes(adata,\n",
    "                           mode=\"int\",\n",
    "                           abs=True,\n",
    "                           output_values=True)\n",
    "    \n",
    "    for i in range(adata.uns['int'].shape[0]):\n",
    "        eff, reg = np.array(adata.uns['int'])[i,0].split(\" - \")\n",
    "        file_name = \"Plots/TF_regulation/Dose\" + dose + \"/\" + reg + \"/\" + reg + \"_\" + eff + \"_umap_\"\n",
    "        print(eff, reg)\n",
    "        dyn.pl.jacobian(adata, regulators=[reg], effectors=[eff], basis='umap', save_show_or_return='save', save_kwargs = {\"path\":file_name, \"ext\":'png'})\n",
    "        \n",
    "    for i in range(adata.uns['int'].shape[0]):\n",
    "        eff, reg = np.array(adata.uns['int'])[i,0].split(\" - \")\n",
    "        file_name = \"Plots/TF_regulation/Dose\" + dose + \"/\" + reg + \"/\" + reg + \"_\" + eff + \"_xy_\"\n",
    "        print(eff, reg)\n",
    "        dyn.pl.jacobian(adata, regulators=[reg], effectors=[eff], x=reg, y=eff, layer='M_s', basis='umap', save_show_or_return='save', save_kwargs = {\"path\":file_name, \"ext\":'png'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "aae7e4a1",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_gene_perturbation(adata, perturb_gene, eff, exp):\n",
    "    dose = np.unique(adata.obs['dosage'])[0]\n",
    "    dyn.pd.perturbation(adata, genes=[perturb_gene], expression = exp, perturb_mode=\"raw\")\n",
    "    dV = vector_transformation(adata.obsm['j_delta_x_perturbation'], adata.uns['PCs'])\n",
    "\n",
    "    all_genes = adata.var_names[adata.var['use_for_pca']]\n",
    "    gene_index = None\n",
    "    if eff in all_genes:\n",
    "        gene_index = all_genes.get_loc(eff)\n",
    "\n",
    "    dv = dV[:,gene_index]\n",
    "    adata.obs['dv_' + eff] = dv\n",
    "    if exp < 0:\n",
    "        file_name = \"Plots/Dose12.5-regulation/\"+perturb_gene+\"/neg_pert_dv_\" + eff \n",
    "    else:\n",
    "        file_name = \"Plots/Dose12.5-regulation/\"+perturb_gene+\"/pos_pert_dv_\" + eff \n",
    "        \n",
    "    #plt.scatter(adata.obsm['X_umap'][:,0], adata.obsm['X_umap'][:,1], c=adata.obs['dv_' + eff], alpha=0.5, cmap='viridis')\n",
    "    plt.scatter(adata.obsm['X_umap'][:,0], adata.obsm['X_umap'][:,1], norm=DivergingNorm(0), c=adata.obs['dv_' + eff], alpha=0.5, cmap='bwr')\n",
    "    plt.title(\"Dose \" + dose + \": \" + perturb_gene + \" peturbation\")\n",
    "    plt.xlabel(\"dv_\" + eff)\n",
    "    plt.colorbar(orientation=\"vertical\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(file_name, dpi=300)\n",
    "    plt.show()\n",
    "    \n",
    "    \n",
    "    dyn.pl.streamline_plot(adata, basis='umap_perturbation', color='dv_' + eff)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c904828",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Scoring function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0298da19",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Calculates the EMT score using correlation coefficients to e-cadherin \n",
    "# gene_exp = raw data, M_s, log transformed gene expression\n",
    "# Adds EMT score and final EMT score to obs of the adata\n",
    "\n",
    "def calc_EMT_score(adata, gene_coeff, gene_exp, threshold):\n",
    "    EMT_score = []\n",
    "    all_genes = adata.var_names\n",
    "    \n",
    "    for i in range(gene_exp.shape[0]):\n",
    "        score = 0\n",
    "        for j in range(gene_coeff.shape[0]):\n",
    "            gene = gene_coeff['GENE'][j] \n",
    "            if gene in adata.var_names.tolist():\n",
    "                gene_index = None\n",
    "                if gene in all_genes:\n",
    "                    gene_index = all_genes.get_loc(gene)\n",
    "                    score += gene_coeff['E-cad'][j]*gene_exp[i,gene_index]\n",
    "        EMT_score.append(score)\n",
    "        \n",
    "    adata.obs['EMT_score'] = EMT_score\n",
    "\n",
    "    final_EMT_score = []\n",
    "    for i in range(len(EMT_score)):\n",
    "        if EMT_score[i] < threshold:\n",
    "            final_EMT_score.append('M')\n",
    "        else:\n",
    "            final_EMT_score.append('E')\n",
    "    adata.obs['final_EMT_score'] = final_EMT_score\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "582f0218",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Calculates the EMT score using correlation coefficients to e-cadherin \n",
    "# gene_exp = raw data, M_s, log transformed gene expression\n",
    "# Adds EMT score and final EMT score to obs of the adata\n",
    "\n",
    "def calc_EMT_score2(adata, gene_coeff, gene_exp, threshold):\n",
    "    EMT_score = []\n",
    "    all_genes = adata.var_names\n",
    "    \n",
    "    for i in range(gene_exp.shape[0]):\n",
    "        score = 0\n",
    "        for j in range(gene_coeff.shape[0]):\n",
    "            gene = gene_coeff['GENE'][j] \n",
    "            #print(gene)\n",
    "            #if gene in adata.var_names.tolist():\n",
    "            #    gene_index = None\n",
    "            if gene in all_genes:\n",
    "                gene_index = all_genes.get_loc(gene)\n",
    "                score += gene_coeff['E-cad'][j]*gene_exp[gene][i]\n",
    "        EMT_score.append(score)\n",
    "        \n",
    "    #adata.obs['EMT_score'] = EMT_score\n",
    "\n",
    "    final_EMT_score = []\n",
    "    for i in range(len(EMT_score)):\n",
    "        if EMT_score[i] < threshold:\n",
    "            final_EMT_score.append('M')\n",
    "        else:\n",
    "            final_EMT_score.append('E')\n",
    "    #adata.obs['final_EMT_score'] = final_EMT_score\n",
    "    return EMT_score, final_EMT_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8e75c11a",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def normalize_expr(expr_dict):\n",
    "    ret = {}\n",
    "    for k, v in expr_dict.items():      \n",
    "        if np.min(v)==0 and np.max(v)==0:\n",
    "            ret[k] = v\n",
    "        else:\n",
    "            ret[k] = (v - np.mean(v)) / (np.max(v) - np.min(v))\n",
    "    return ret\n",
    "\n",
    "def emt_score(norm_expr_dict, k):\n",
    "    ret = np.zeros(k)\n",
    "    for i in range(k):\n",
    "        ret[i] = norm_expr_dict['COL1A2'][i] + norm_expr_dict['FN1'][i] + norm_expr_dict['SNAI1'][i] \\\n",
    "            + norm_expr_dict['SNAI2'][i] \\\n",
    "            + norm_expr_dict['VIM'][i] + norm_expr_dict['ZEB1'][i] \\\n",
    "            - norm_expr_dict['ERBB3'][i] - norm_expr_dict['KRT15'][i] - norm_expr_dict['PSCA'][i] - norm_expr_dict['MAL2'][i] \\\n",
    "            - norm_expr_dict['S100P'][i]\n",
    "    return ret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30170611",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def KS_emt_score(E_exp, M_exp):\n",
    "    finalScore = []\n",
    "    for i in range(E_exp.shape[0]):\n",
    "        #stat, pval = ks_2samp(E_exp.iloc[i,:], M_exp.iloc[i,:])\n",
    "        stat2, pval2 = ks_2samp(E_exp.iloc[i,:], M_exp.iloc[i,:], alternative='less')\n",
    "        stat3, pval3 = ks_2samp(E_exp.iloc[i,:], M_exp.iloc[i,:], alternative='greater')\n",
    "\n",
    "        if pval2 < 0.05:\n",
    "            finalScore.append(-1*stat2)\n",
    "        elif pval3 < 0.05:\n",
    "            finalScore.append(stat3)\n",
    "        else:\n",
    "            if stat3 > stat2:\n",
    "                finalScore.append(stat3)\n",
    "            else:\n",
    "                finalScore.append(-1*stat2)\n",
    "    return finalScore"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cb0245b",
   "metadata": {
    "heading_collapsed": true,
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "03338712",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# converts pM to ng/ml\n",
    "# 1 ng/ml = 40 pM\n",
    "def convert_pM(nums):\n",
    "    toReturn = []\n",
    "    for i in range(len(nums)):\n",
    "        toReturn.append(nums[i]/40.0)\n",
    "    return toReturn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "187bf922",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# converts ng/ml to pM\n",
    "# 1 ng/ml = 40 pM\n",
    "def convert_ngml(nums):\n",
    "    toReturn = []\n",
    "    for i in range(len(nums)):\n",
    "        toReturn.append(nums[i]*40.0)\n",
    "    return toReturn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3f8ac000",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Prints whether or not gene is in given gene list\n",
    "def findGene(gene_name, gene_list):\n",
    "    inSet = gene_name in gene_list\n",
    "    if inSet == True:\n",
    "        print(gene_name, \"\\t True\")\n",
    "    else:\n",
    "        print(gene_name, \"\\t False\")\n",
    "    return inSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccf39f47",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def getLOG2_FoldChange(expr):\n",
    "    toReturn = [0]\n",
    "    control = expr[0]\n",
    "    for i in range(1,len(expr)):\n",
    "        toReturn.append(np.log2(expr[i]/control))\n",
    "    return toReturn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "94579d22",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def getFoldChange(expr):\n",
    "    toReturn = [0]\n",
    "    control = None\n",
    "    if expr[0] == 0:\n",
    "        control = 0.001\n",
    "    else:\n",
    "        control = expr[0]\n",
    "    for i in range(1,len(expr)):\n",
    "        toReturn.append((expr[i] - control)/control)\n",
    "    return toReturn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d50c5dcc",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_expression(gene, dose_cells, process_adata2, folder):\n",
    "    intervals = np.linspace(-1,1,4)\n",
    "    exp_mean = []\n",
    "    exp_std = []\n",
    "    exp = []\n",
    "    my_cmap = plt.get_cmap(\"viridis\")\n",
    "    y = range(8)\n",
    "    rescale = lambda y: (y - np.min(y)) / (np.max(y) - np.min(y))\n",
    "    plt.figure(figsize=(5,3.5), dpi=100)\n",
    "\n",
    "    interval_top = None\n",
    "    interval_bottom = None\n",
    "\n",
    "    group_index = [0,1,1,1,2,2,2,2]\n",
    "    for i in range(len(dose_cells)):\n",
    "        if i == 0:\n",
    "            interval_top = intervals[group_index[i]+2]\n",
    "            interval_bottom = intervals[group_index[i]]\n",
    "        else:\n",
    "            interval_top = intervals[group_index[i]+1]\n",
    "            interval_bottom = intervals[group_index[i]]\n",
    "\n",
    "        adata_temp = processed_adata2[np.where(processed_adata2.obs['dosage'] == str(doses[i]))[0],:]\n",
    "        indexes = np.where((adata_temp.obs['emt_score'] < interval_top) & (adata_temp.obs['emt_score'] >= interval_bottom))[0]\n",
    "        gene_names = adata_temp.var.index\n",
    "        gene_index = gene_names.get_loc(gene)\n",
    "        expr = adata_temp.X.toarray()[indexes, gene_index]\n",
    "        exp_mean.append(np.mean(expr))\n",
    "        exp_std.append(np.std(expr))\n",
    "        exp.append(expr)\n",
    "\n",
    "    fc = getLOG2_FoldChange(exp_mean)\n",
    "\n",
    "    plt.bar(range(8), exp_mean, color=my_cmap(rescale(y)))\n",
    "    plt.xticks([0,1,2,3,4,5,6,7], doses)\n",
    "    plt.xlabel(\"TGFB dose\")\n",
    "    plt.ylabel(\"Gene Expression\")\n",
    "    plt.title(gene + \" expression across TGFB doses\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"Plots/TF_expression/\" + folder +\"/\"+ gene + \"_ge_3_selected_groups.png\", dpi=200)\n",
    "    plt.show()\n",
    "\n",
    "    plt.figure(figsize=(5,3.5), dpi=100)\n",
    "    plt.bar(range(7), fc, color=my_cmap(rescale(y)[1:]))\n",
    "    plt.xticks([0,1,2,3,4,5,6], doses[1:])\n",
    "    plt.xlabel(\"TGFB dose\")\n",
    "    plt.ylabel(\"log2 fold change\")\n",
    "    plt.title(gene + \" expression across TGFB doses\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"Plots/TF_expression/\"+folder+\"/\"+gene+\"_fc_3_selected_groups.png\", dpi=200)\n",
    "    plt.show()\n",
    "\n",
    "    plt.figure(figsize=(5,3.5), dpi=100)\n",
    "    plt.bar(range(7), fc, color=my_cmap(rescale(y)[1:]))\n",
    "    plt.xticks([0,1,2,3,4,5,6], converted_units[1:])\n",
    "    plt.xlabel(\"TGFB dose\")\n",
    "    plt.ylabel(\"log2 fold change\")\n",
    "    plt.title(gene + \" expression across TGFB doses\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"Plots/TF_expression/\"+folder+\"/\"+gene+\"_fc_3_selected_groups_JZ_units.png\", dpi=200)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55ecda96",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def plot_expression_A549(gene, adata_timepoints, folder):\n",
    "    exp_mean = []\n",
    "    exp = []\n",
    "    my_cmap = plt.get_cmap(\"viridis\")\n",
    "    y = range(4)\n",
    "    rescale = lambda y: (y - np.min(y)) / (np.max(y) - np.min(y))\n",
    "    plt.figure(figsize=(5,3.5), dpi=100)\n",
    "\n",
    "    for i in range(len(timepoints)):\n",
    "        adata_temp = adata_timepoints[np.where(adata_timepoints.obs['treatment'] == timepoints[i])[0],:]\n",
    "        gene_names = adata_temp.var.index\n",
    "        gene_index = gene_names.get_loc(gene)\n",
    "        expr = adata_temp.X.toarray()[:, gene_index]\n",
    "        exp_mean.append(np.mean(expr))\n",
    "        exp.append(expr)\n",
    "\n",
    "    fc = getLOG2_FoldChange(exp_mean)  \n",
    "\n",
    "    plt.bar(range(4), exp_mean, color=my_cmap(rescale(y)))\n",
    "    plt.xticks([0,1,2,3], timepoints)\n",
    "    plt.xlabel(\"Timepoint\")\n",
    "    plt.ylabel(\"Gene Expression\")\n",
    "    plt.title(gene + \"expression with time\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"../Plots/TF_expression/A549/\"+folder+\"/\"+gene+\"_ge.png\", dpi=200)\n",
    "    plt.show()\n",
    "\n",
    "    plt.figure(figsize=(5,3.5), dpi=100)\n",
    "    plt.bar(range(3), fc, color=my_cmap(rescale(y)[1:]))\n",
    "    plt.xticks([0,1,2], timepoints[1:])\n",
    "    plt.xlabel(\"Timepoint\")\n",
    "    plt.ylabel(\"log2 fold change\")\n",
    "    plt.title(gene + \"expression with time\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"../Plots/TF_expression/A549/\"+folder+\"/\"+gene+\"_fc.png\", dpi=200)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa5463ee",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def drem2_normalize(expr):\n",
    "    for i in range(expr.shape[0]):\n",
    "        start = expr[i][0]\n",
    "        expr[i][1:] = np.log2(expr[i][1:]/start)\n",
    "    return expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3597a712",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def gseapy_barplot(df, column='Adjusted P-value', title=\"\", cutoff=0.05, top_term=10,\n",
    "            figsize=(6.5,6), color='salmon', ofname=None, **kwargs):\n",
    "    \"\"\"Visualize enrichr results.\n",
    "    :param df: GSEApy DataFrame results.\n",
    "    :param column: which column of DataFrame to show. Default: Adjusted P-value\n",
    "    :param title: figure title.\n",
    "    :param cutoff: terms with 'column' value < cut-off are shown.\n",
    "    :param top_term: number of top enriched terms to show.\n",
    "    :param figsize: tuple, matplotlib figsize.\n",
    "    :param color: color for bars.\n",
    "    :param ofname: output file name. If None, don't save figure    \n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    colname = column   \n",
    "    if colname in ['Adjusted P-value', 'P-value']: \n",
    "        # check if any values in `df[colname]` can't be coerced to floats\n",
    "        can_be_coerced = df[colname].map(isfloat)\n",
    "        if np.sum(~can_be_coerced) > 0:\n",
    "            raise ValueError('some value in %s could not be typecast to `float`'%colname)\n",
    "        else:\n",
    "            df.loc[:, colname] = df[colname].map(float)\n",
    "        df = df[df[colname] <= cutoff]\n",
    "        if len(df) < 1: \n",
    "            msg = \"Warning: No enrich terms using library %s when cutoff = %s\"%(title, cutoff)\n",
    "            return msg\n",
    "        df = df.assign(logAP = lambda x: - x[colname].apply(np.log10))\n",
    "        colname = 'logAP' \n",
    "        \n",
    "    dd = df.sort_values(by=colname).iloc[-top_term:,:]\n",
    "    # create bar plot\n",
    "    if hasattr(sys, 'ps1') and (ofname is None):\n",
    "        # working inside python console, show (True) figure\n",
    "        fig = plt.figure(figsize=figsize)\n",
    "    else:\n",
    "        # If working on commandline, don't show figure\n",
    "        fig = Figure(figsize=figsize)\n",
    "        canvas = FigureCanvas(fig)\n",
    "    ax = fig.add_subplot(111)\n",
    "    bar = dd.plot.barh(x='Term', y=colname, color=color, \n",
    "                       alpha=0.75, fontsize=9, ax=ax)\n",
    "    \n",
    "    if column in ['Adjusted P-value', 'P-value']:\n",
    "        xlabel = \"-log$_{10}$(%s)\"%column\n",
    "    else:\n",
    "        xlabel = column \n",
    "    bar.set_xlabel(xlabel, fontsize=9, fontweight='bold')\n",
    "    bar.set_ylabel(\"\")\n",
    "    bar.set_title(title, fontsize=12, fontweight='bold')\n",
    "    bar.xaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "    bar.legend_.remove()\n",
    "    adjust_spines(ax, spines=['left','bottom'])\n",
    "\n",
    "    if hasattr(sys, 'ps1') and (ofname is not None): \n",
    "        # canvas.print_figure(ofname, bbox_inches='tight', dpi=300)\n",
    "        fig.savefig(ofname, bbox_inches='tight', dpi=300)\n",
    "        return\n",
    "    elif ofname is None:\n",
    "        return\n",
    "    elif not hasattr(sys, 'ps1') and (ofname is not None):\n",
    "        canvas.print_figure(ofname, bbox_inches='tight', dpi=300)\n",
    "        return\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3acac8ca",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Plots the gene expression for a given basis\n",
    "def plotGeneExpression(adata, gene, basis):\n",
    "    gene_index = adata.var.index.get_loc(gene)\n",
    "    exp = adata.X.toarray()[:,gene_index]\n",
    "    temp_indices = np.where(exp==0)\n",
    "    plt.scatter(adata.obsm['X_'+basis][temp_indices,0], adata.obsm['X_'+basis][temp_indices,1], c='gainsboro', s=75)\n",
    "    temp_indices = np.where(exp!=0)\n",
    "#     print(adata.X.toarray()[temp_indices,gene_index])\n",
    "#     print(adata.obsm['X_'+basis][temp_indices,0])\n",
    "#     print(adata.obsm['X_'+basis][temp_indices,1])\n",
    "    plt.scatter(adata.obsm['X_'+basis][temp_indices,0], adata.obsm['X_'+basis][temp_indices,1], c=adata.X.toarray()[temp_indices,gene_index][0], s=75, vmin=0)\n",
    "    plt.title(gene)\n",
    "    plt.colorbar()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77b95ce9",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b95c7bdb",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a05f868d",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a0243c9",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7a473f27",
   "metadata": {
    "heading_collapsed": true,
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Weikang's functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b7af595e",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "#https://gist.github.com/pv/8036995\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.spatial import Voronoi\n",
    "\n",
    "def voronoi_finite_polygons_2d(vor, radius=None):\n",
    "    \"\"\"\n",
    "    Reconstruct infinite voronoi regions in a 2D diagram to finite\n",
    "    regions.\n",
    "    Parameters\n",
    "    ----------\n",
    "    vor : Voronoi\n",
    "        Input diagram\n",
    "    radius : float, optional\n",
    "        Distance to 'points at infinity'.\n",
    "    Returns\n",
    "    -------\n",
    "    regions : list of tuples\n",
    "        Indices of vertices in each revised Voronoi regions.\n",
    "    vertices : list of tuples\n",
    "        Coordinates for revised Voronoi vertices. Same as coordinates\n",
    "        of input vertices, with 'points at infinity' appended to the\n",
    "        end.\n",
    "    \"\"\"\n",
    "\n",
    "    if vor.points.shape[1] != 2:\n",
    "        raise ValueError(\"Requires 2D input\")\n",
    "\n",
    "    new_regions = []\n",
    "    new_vertices = vor.vertices.tolist()\n",
    "\n",
    "    center = vor.points.mean(axis=0)\n",
    "    if radius is None:\n",
    "        radius = vor.points.ptp().max()*2\n",
    "\n",
    "    # Construct a map containing all ridges for a given point\n",
    "    all_ridges = {}\n",
    "    for (p1, p2), (v1, v2) in zip(vor.ridge_points, vor.ridge_vertices):\n",
    "        all_ridges.setdefault(p1, []).append((p2, v1, v2))\n",
    "        all_ridges.setdefault(p2, []).append((p1, v1, v2))\n",
    "\n",
    "    # Reconstruct infinite regions\n",
    "    for p1, region in enumerate(vor.point_region):\n",
    "        vertices = vor.regions[region]\n",
    "\n",
    "        if all(v >= 0 for v in vertices):\n",
    "            # finite region\n",
    "            new_regions.append(vertices)\n",
    "            continue\n",
    "\n",
    "        # reconstruct a non-finite region\n",
    "        ridges = all_ridges[p1]\n",
    "        new_region = [v for v in vertices if v >= 0]\n",
    "\n",
    "        for p2, v1, v2 in ridges:\n",
    "            if v2 < 0:\n",
    "                v1, v2 = v2, v1\n",
    "            if v1 >= 0:\n",
    "                # finite ridge: already in the region\n",
    "                continue\n",
    "\n",
    "            # Compute the missing endpoint of an infinite ridge\n",
    "\n",
    "            t = vor.points[p2] - vor.points[p1] # tangent\n",
    "            t /= np.linalg.norm(t)\n",
    "            n = np.array([-t[1], t[0]])  # normal\n",
    "\n",
    "            midpoint = vor.points[[p1, p2]].mean(axis=0)\n",
    "            direction = np.sign(np.dot(midpoint - center, n)) * n\n",
    "            far_point = vor.vertices[v2] + direction * radius\n",
    "\n",
    "            new_region.append(len(new_vertices))\n",
    "            new_vertices.append(far_point.tolist())\n",
    "\n",
    "        # sort region counterclockwise\n",
    "        vs = np.asarray([new_vertices[v] for v in new_region])\n",
    "        c = vs.mean(axis=0)\n",
    "        angles = np.arctan2(vs[:,1] - c[1], vs[:,0] - c[0])\n",
    "        new_region = np.array(new_region)[np.argsort(angles)]\n",
    "\n",
    "        # finish\n",
    "        new_regions.append(new_region.tolist())\n",
    "\n",
    "    return new_regions, np.asarray(new_vertices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "854f5936",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def is_outlier(points, thresh=3.5):\n",
    "    \"\"\"\n",
    "    Returns a boolean array with True if points are outliers and False \n",
    "    otherwise.\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "        points : An numobservations by numdimensions array of observations\n",
    "        thresh : The modified z-score to use as a threshold. Observations with\n",
    "            a modified z-score (based on the median absolute deviation) greater\n",
    "            than this value will be classified as outliers.\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "        mask : A numobservations-length boolean array.\n",
    "\n",
    "    References:\n",
    "    ----------\n",
    "        Boris Iglewicz and David Hoaglin (1993), \"Volume 16: How to Detect and\n",
    "        Handle Outliers\", The ASQC Basic References in Quality Control:\n",
    "        Statistical Techniques, Edward F. Mykytka, Ph.D., Editor. \n",
    "    \"\"\"\n",
    "    if len(points.shape) == 1:\n",
    "        points = points[:,None]\n",
    "    median = np.median(points, axis=0)\n",
    "    diff = np.sum((points - median)**2, axis=-1)\n",
    "    diff = np.sqrt(diff)\n",
    "    med_abs_deviation = np.median(diff)\n",
    "\n",
    "    modified_z_score = 0.6745 * diff / med_abs_deviation\n",
    "\n",
    "    return modified_z_score > thresh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c22ac521",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def get_directed_edges(trans_matrix):\n",
    "    edges={}\n",
    "    for i in range(trans_matrix.shape[0]):\n",
    "        for j in range(trans_matrix.shape[1]):\n",
    "            if i!=j and trans_matrix[i,j]>1e-2:\n",
    "                edges[(i,j)]=-np.log(trans_matrix[i,j])\n",
    "    return edges\n",
    "def get_undirected_edges(trans_matrix):\n",
    "        edges={}\n",
    "        for i in range(trans_matrix.shape[0]):\n",
    "            for j in range(i+1,trans_matrix.shape[1]):\n",
    "                if trans_matrix[i,j]+trans_matrix[j,i]>2e-2:\n",
    "                    edges[(i,j)]=trans_matrix[i,j]+trans_matrix[j,i]\n",
    "        return edges\n",
    "\n",
    "# def get_directed_edges(trans_matrix):\n",
    "#     conn_indices = np.where(trans_matrix)\n",
    "#     weights = trans_matrix[conn_indices]\n",
    "#     edges = zip(*conn_indices)\n",
    "#     edge_wts = dict(zip(edges, weights))\n",
    "#     return edge_wts\n",
    " \n",
    "def build_trans_networkx(trans_matrix,node_list,direct_flag=True):\n",
    "    \n",
    "    if direct_flag:\n",
    "        G=nx.DiGraph()\n",
    "        G.add_nodes_from(node_list)    \n",
    "        edges_wts=get_directed_edges(trans_matrix)\n",
    "    # edges represent transition probabilities\n",
    "        for k, v in edges_wts.items():\n",
    "            tmp_origin, tmp_destination = k[0], k[1]\n",
    "            G.add_edge(tmp_origin, tmp_destination, weight=v, label=v)\n",
    "    else:\n",
    "        G=nx.Graph()\n",
    "        G.add_nodes_from(node_list)    \n",
    "        edges_wts=get_undirected_edges(trans_matrix)\n",
    "        # edges represent transition probabilities\n",
    "        for k, v in edges_wts.items():\n",
    "            tmp_origin, tmp_destination = k[0], k[1]\n",
    "            G.add_edge(tmp_origin, tmp_destination, weight=v, label=v)\n",
    "        \n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b618b8d8-f3ba-4fab-b5b9-36c2f3bd0253",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "from scipy.interpolate import splprep,splev\n",
    "from sklearn.cluster import DBSCAN\n",
    "from scipy import interpolate\n",
    "from numpy import linalg \n",
    "\"\"\"\n",
    "Find path connecting data point\n",
    "\"\"\"\n",
    "class FindPath2(object):\n",
    "    def __init__(self, n = 20,traj_w=10,sf=1, ):\n",
    "        self.n = n\n",
    "        self.t = np.linspace(0,1,n)\n",
    "        self.error = 1e100\n",
    "        self.error_log=[]\n",
    "        self.data_rc=[]\n",
    "        self.tangent_v=[]\n",
    "        self.w=traj_w\n",
    "        self.sf=sf\n",
    "        #self.coords = coords\n",
    "        #self.spline = spline\n",
    "        #self.path = path\n",
    "        \n",
    "    def fit(self, p1, p2, data):\n",
    "        self.p1 = np.expand_dims(p1,axis=0)\n",
    "        self.p2 = np.expand_dims(p2,axis=0)\n",
    "        self.data = data\n",
    "        self.cluster_labels=[]\n",
    "\n",
    "        self.spline = p1 + np.multiply((p2 - p1) , np.expand_dims(self.t,axis=1))\n",
    "        #print(self.spline)\n",
    "        # plt.scatter(self.coords['DC1'],\n",
    "        #             self.coords['DC2'],\n",
    "        #             c='gainsboro', zorder=1)\n",
    "        plt.plot(self.spline[:,0], self.spline[:,1], zorder=2) \n",
    "        plt.show()\n",
    "        \n",
    "        self.best_spline=self.spline.copy()\n",
    "        \n",
    "        update=True\n",
    "        while update:\n",
    "            update=self.__move()\n",
    "            print(update)\n",
    "        return self.best_spline,self.data_rc\n",
    "    \n",
    "    def __move(self):\n",
    "        w=self.w\n",
    "        traj_error_arr1,traj_sum1,pts_error_arr1,pts_sum1,density1,data_rc1=self.__error() \n",
    "        centers = self.spline.copy()\n",
    "#         print(pts_sum1) \n",
    "        \n",
    "        updateIndexs = np.where(density1>0)[0]\n",
    "        centers[updateIndexs,:]=(np.divide(pts_sum1[updateIndexs,:],np.expand_dims(density1[updateIndexs],axis=1))+\\\n",
    "                                w*traj_sum1[updateIndexs,:]/len(self.data))/(1+w)\n",
    "\n",
    "\n",
    "        self.spline,tangent_v=self.__discretize(centers)\n",
    "        \n",
    "        dot_color=np.arange(self.spline.shape[0])\n",
    "        cm=plt.cm.get_cmap('jet')\n",
    "        plt.figure(figsize=(7.5,5))\n",
    "        # plt.scatter(self.coords['DC1'],\n",
    "        #             self.coords['DC2'],\n",
    "        #             c='gainsboro', zorder=1)\n",
    "        plt.plot(self.spline[:,0],self.spline[:,1], linestyle='dotted', c='black', zorder=2)\n",
    "        plt.scatter(self.spline[:,0],self.spline[:,1],c=dot_color,cmap=cm, zorder=3)\n",
    "        plt.xlabel('X',fontsize=16)\n",
    "        plt.ylabel('Y',fontsize=16)\n",
    "        plt.xticks(fontsize=14)\n",
    "        plt.yticks(fontsize=14)\n",
    "        #plt.savefig(self.path + \"/\" + str(self.error)+'.png',dpi=300)\n",
    "        plt.show()\n",
    "#         print(tangent_v)\n",
    "        traj_error_arr2,traj_sum2,pts_error_arr2,pts_sum2,density2,data_rc2=self.__error() \n",
    "\n",
    "#         print(w*np.sum(traj_error_arr2),np.sum(pts_error_arr2))\n",
    "        error=w*np.sum(traj_error_arr2)+np.sum(pts_error_arr2)\n",
    "#         print(self.error_log)\n",
    "\n",
    "        self.error_log.append(error)\n",
    "\n",
    "        if error<self.error:\n",
    "            self.error=error\n",
    "            self.best_spline=self.spline.copy()\n",
    "            self.data_rc=data_rc2\n",
    "            self.tangent_v=tangent_v\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "\n",
    "    def __discretize(self,centers,sample_n=1,st_move=0.02):\n",
    "        #centers[0,:] = self.p1\n",
    "        #centers[-1,:] = self.p2\n",
    "        spline=centers.copy()\n",
    "        centers[0,:]=self.p1#+st_move*(centers[0,:]-self.p1)#self.spline[0,:]+st_move*(centers[0,:]-self.spline[0,:])\n",
    "        centers[-1,:]=self.p2#+st_move*(centers[-1,:]-self.p2)#self.spline[-1,:]+st_move*(centers[-1,:]-self.spline[-1,:])\n",
    "        t = np.linspace(0,1, sample_n* self.n)\n",
    "\n",
    "\n",
    "        center_list=[centers[:,i] for i in range(centers.shape[1])]\n",
    "        tck, u = interpolate.splprep(center_list,s=self.sf,k=2)\n",
    "        spline_list=splev(t,tck,ext=3)\n",
    "        for k in range(len(spline_list)):\n",
    "            spline[:,k]=spline_list[k][::sample_n]\n",
    "        \n",
    "        \n",
    "#         plt.title('smooth')\n",
    "#         plt.plot(self.spline[:,0], self.spline[:,1]) \n",
    "#         plt.show()\n",
    "#         plt.plot(self.spline[:,2], self.spline[:,3]) \n",
    "#         plt.show()\n",
    "        \n",
    "        tangent_v=interpolate.splev(t,tck,der=1)\n",
    "        \n",
    "        return spline,tangent_v\n",
    "        \n",
    "    def __error(self):\n",
    "          \n",
    "        pts_error_arr=np.zeros(self.n)\n",
    "        density = np.zeros(self.n)\n",
    "        pts_sum=np.zeros_like(self.spline)\n",
    "        \n",
    "        traj_error_arr=np.zeros(self.n)\n",
    "        traj_sum=np.zeros_like(self.spline)\n",
    "        data_rc=[]\n",
    "        for ind in range(len(self.data)):\n",
    "            traj_t_span = self.data[ind].shape[0]\n",
    "            spline_inds=[]\n",
    "            spline_dist=[]\n",
    "            for j in range(self.n):\n",
    "                p=self.spline[j]\n",
    "                #print(self.data[ind])\n",
    "                #print(p[None,:])\n",
    "                dist=linalg.norm(self.data[ind]-p[None,:],axis=1)\n",
    "#                 print(dist.shape)\n",
    "                traj_error_arr[j]+=np.amin(dist)\n",
    "                traj_sum[j]+=self.data[ind][np.argmin(dist),:]\n",
    "                \n",
    "\n",
    "            traj_rc=np.zeros((self.data[ind].shape[0],))\n",
    "            for i in range(traj_t_span):\n",
    "                p = self.data[ind][i,:]\n",
    "                dist = linalg.norm(self.spline - p[None,:], axis = 1)\n",
    "                pts_error_arr[np.argmin(dist)]+=np.amin(dist)\n",
    "                pts_sum[np.argmin(dist)]+=self.data[ind][i,:]\n",
    "                density[np.argmin(dist)]+=1\n",
    "                traj_rc[i]=np.argmin(dist)\n",
    "            data_rc.append(traj_rc)\n",
    "        \n",
    "        return traj_error_arr,traj_sum,pts_error_arr,pts_sum,density,data_rc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4c67a366",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "from scipy.interpolate import splprep,splev\n",
    "from sklearn.cluster import DBSCAN\n",
    "from scipy import interpolate\n",
    "from numpy import linalg \n",
    "\"\"\"\n",
    "Find path connecting data point\n",
    "\"\"\"\n",
    "class FindPath(object):\n",
    "    def __init__(self, coords, spline, path, n = 20,traj_w=10,sf=1, ):\n",
    "        self.n = n\n",
    "        self.t = np.linspace(0,1,n)\n",
    "        self.error = 1e100\n",
    "        self.error_log=[]\n",
    "        self.data_rc=[]\n",
    "        self.tangent_v=[]\n",
    "        self.w=traj_w\n",
    "        self.sf=sf\n",
    "        self.coords = coords\n",
    "        self.spline = spline\n",
    "        self.path = path\n",
    "        \n",
    "    def fit(self, p1, p2, data):\n",
    "        self.p1 = np.expand_dims(p1,axis=0)\n",
    "        self.p2 = np.expand_dims(p2,axis=0)\n",
    "        self.data = data\n",
    "        self.cluster_labels=[]\n",
    "\n",
    "        #self.spline = p1 + np.multiply((p2 - p1) , np.expand_dims(self.t,axis=1))\n",
    "        plt.scatter(self.coords['DC1'],\n",
    "                    self.coords['DC2'],\n",
    "                    c='gainsboro', zorder=1)\n",
    "        plt.plot(self.spline[:,0], self.spline[:,1], zorder=2) \n",
    "        plt.show()\n",
    "        \n",
    "        self.best_spline=self.spline.copy()\n",
    "        \n",
    "        update=True\n",
    "        while update:\n",
    "            update=self.__move()\n",
    "            print(update)\n",
    "        return self.best_spline,self.data_rc\n",
    "    \n",
    "    def __move(self):\n",
    "        w=self.w\n",
    "        traj_error_arr1,traj_sum1,pts_error_arr1,pts_sum1,density1,data_rc1=self.__error() \n",
    "        centers = self.spline.copy()\n",
    "#         print(pts_sum1) \n",
    "        \n",
    "        updateIndexs = np.where(density1>0)[0]\n",
    "        centers[updateIndexs,:]=(np.divide(pts_sum1[updateIndexs,:],np.expand_dims(density1[updateIndexs],axis=1))+\\\n",
    "                                w*traj_sum1[updateIndexs,:]/len(self.data))/(1+w)\n",
    "\n",
    "\n",
    "        self.spline,tangent_v=self.__discretize(centers)\n",
    "        \n",
    "        dot_color=np.arange(self.spline.shape[0])\n",
    "        cm=plt.cm.get_cmap('jet')\n",
    "        plt.figure(figsize=(7.5,5))\n",
    "        plt.scatter(self.coords['DC1'],\n",
    "                    self.coords['DC2'],\n",
    "                    c='gainsboro', zorder=1)\n",
    "        plt.plot(self.spline[:,0],self.spline[:,1], linestyle='dotted', c='black', zorder=2)\n",
    "        plt.scatter(self.spline[:,0],self.spline[:,1],c=dot_color,cmap=cm, zorder=3)\n",
    "        plt.xlabel('X',fontsize=16)\n",
    "        plt.ylabel('Y',fontsize=16)\n",
    "        plt.xticks(fontsize=14)\n",
    "        plt.yticks(fontsize=14)\n",
    "        plt.savefig(self.path + \"/\" + str(self.error)+'.png',dpi=300)\n",
    "        plt.show()\n",
    "#         print(tangent_v)\n",
    "        traj_error_arr2,traj_sum2,pts_error_arr2,pts_sum2,density2,data_rc2=self.__error() \n",
    "\n",
    "#         print(w*np.sum(traj_error_arr2),np.sum(pts_error_arr2))\n",
    "        error=w*np.sum(traj_error_arr2)+np.sum(pts_error_arr2)\n",
    "#         print(self.error_log)\n",
    "\n",
    "        self.error_log.append(error)\n",
    "\n",
    "        if error<self.error:\n",
    "            self.error=error\n",
    "            self.best_spline=self.spline.copy()\n",
    "            self.data_rc=data_rc2\n",
    "            self.tangent_v=tangent_v\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "\n",
    "    def __discretize(self,centers,sample_n=1,st_move=0.02):\n",
    "        #centers[0,:] = self.p1\n",
    "        #centers[-1,:] = self.p2\n",
    "        spline=centers.copy()\n",
    "        centers[0,:]=self.p1#+st_move*(centers[0,:]-self.p1)#self.spline[0,:]+st_move*(centers[0,:]-self.spline[0,:])\n",
    "        centers[-1,:]=self.p2#+st_move*(centers[-1,:]-self.p2)#self.spline[-1,:]+st_move*(centers[-1,:]-self.spline[-1,:])\n",
    "        t = np.linspace(0,1, sample_n* self.n)\n",
    "\n",
    "\n",
    "        center_list=[centers[:,i] for i in range(centers.shape[1])]\n",
    "        tck, u = interpolate.splprep(center_list,s=self.sf,k=2)\n",
    "        spline_list=splev(t,tck,ext=3)\n",
    "        for k in range(len(spline_list)):\n",
    "            spline[:,k]=spline_list[k][::sample_n]\n",
    "        \n",
    "        \n",
    "#         plt.title('smooth')\n",
    "#         plt.plot(self.spline[:,0], self.spline[:,1]) \n",
    "#         plt.show()\n",
    "#         plt.plot(self.spline[:,2], self.spline[:,3]) \n",
    "#         plt.show()\n",
    "        \n",
    "        tangent_v=interpolate.splev(t,tck,der=1)\n",
    "        \n",
    "        return spline,tangent_v\n",
    "        \n",
    "    def __error(self):\n",
    "          \n",
    "        pts_error_arr=np.zeros(self.n)\n",
    "        density = np.zeros(self.n)\n",
    "        pts_sum=np.zeros_like(self.spline)\n",
    "        \n",
    "        traj_error_arr=np.zeros(self.n)\n",
    "        traj_sum=np.zeros_like(self.spline)\n",
    "        data_rc=[]\n",
    "        for ind in range(len(self.data)):\n",
    "            traj_t_span = self.data[ind].shape[0]\n",
    "            spline_inds=[]\n",
    "            spline_dist=[]\n",
    "            for j in range(self.n):\n",
    "                p=self.spline[j]\n",
    "                \n",
    "                dist=linalg.norm(self.data[ind]-p[None,:],axis=1)\n",
    "#                 print(dist.shape)\n",
    "                traj_error_arr[j]+=np.amin(dist)\n",
    "                traj_sum[j]+=self.data[ind][np.argmin(dist),:]\n",
    "                \n",
    "\n",
    "            traj_rc=np.zeros((self.data[ind].shape[0],))\n",
    "            for i in range(traj_t_span):\n",
    "                p = self.data[ind][i,:]\n",
    "                dist = linalg.norm(self.spline - p[None,:], axis = 1)\n",
    "                pts_error_arr[np.argmin(dist)]+=np.amin(dist)\n",
    "                pts_sum[np.argmin(dist)]+=self.data[ind][i,:]\n",
    "                density[np.argmin(dist)]+=1\n",
    "                traj_rc[i]=np.argmin(dist)\n",
    "            data_rc.append(traj_rc)\n",
    "        \n",
    "        return traj_error_arr,traj_sum,pts_error_arr,pts_sum,density,data_rc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a3ff3d-093b-40be-b1b7-ddab0051c6e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88b4add1-a83d-4f9d-b745-779ec815dcae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6636d1b7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Trajectory simulation functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "279860bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cos_corr(xi, Xj, vi):\n",
    "    D = Xj - xi\n",
    "    dist = np.linalg.norm(D, axis=1)\n",
    "    dist[dist == 0] = 1\n",
    "    D /= dist[:, None]\n",
    "\n",
    "    v_norm = np.linalg.norm(vi)\n",
    "    if v_norm == 0:\n",
    "        v_norm = 1\n",
    "    \n",
    "    return D @ vi / v_norm\n",
    "\n",
    "# X = X or X_spliced/M_s\n",
    "# V = velocity_S\n",
    "# nbrs = uns.neighbors (indices)\n",
    "def corr_kernel(X, V, nbrs, sigma=10, corr_func=cos_corr):\n",
    "    P = np.zeros((X.shape[0], X.shape[0]))\n",
    "    for i, x in enumerate(X):\n",
    "        v, idx = V[i], nbrs[i]\n",
    "\n",
    "        c = corr_func(x, X[idx], v)\n",
    "        p = np.exp(c/sigma)\n",
    "        p /= np.sum(p)\n",
    "        P[i][idx] = p\n",
    "    return P"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c53b03f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def markov_simulation(T, num_trajs, starts, targets):\n",
    "    final_trajs = []\n",
    "    for i in range(num_trajs):\n",
    "        states = [random.choice(starts)]\n",
    "        t=0\n",
    "        while states[t] not in targets:\n",
    "            p = T[states[t],:]\n",
    "            temp = np.where(np.random.multinomial(1, p))[0][0]\n",
    "            if temp in start_pos and len(states)!=1:\n",
    "                states = [temp]\n",
    "                t=0\n",
    "            else:   \n",
    "                states.append(temp)\n",
    "                t+=1\n",
    "        final_trajs.append(states)\n",
    "    return final_trajs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "41719bec-e0af-4fda-b7dd-7beb20e9d8a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def markov_simulation2(T, num_trajs, starts, targets):\n",
    "    final_trajs = []\n",
    "    for i in range(num_trajs):\n",
    "        states = [random.choice(starts)]\n",
    "        t=0\n",
    "        while states[t] not in targets:\n",
    "            p = T[states[t],:]\n",
    "            temp = np.where(np.random.multinomial(1, p))[0][0]\n",
    "            #if temp in starts and len(states)!=1:\n",
    "            if temp in starts:\n",
    "                states = [temp]\n",
    "                t=0\n",
    "            else:   \n",
    "                states.append(temp)\n",
    "                t+=1\n",
    "        final_trajs.append(states)\n",
    "    return final_trajs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e53329d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Returns the distance matrix between LAP trajectories\n",
    "def distanceMatrix(paths):\n",
    "    toReturn = np.zeros((len(paths), len(paths)))\n",
    "    row = 0 \n",
    "    col = 0\n",
    "    while row < toReturn.shape[0]:\n",
    "        while col < toReturn.shape[0]:\n",
    "            dist, _ = fastdtw(paths[row], paths[col])\n",
    "            toReturn[row, col] = dist\n",
    "            toReturn[col, row] = dist\n",
    "            col += 1\n",
    "        col = row+1\n",
    "        row += 1\n",
    "    return toReturn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "62c5a84e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Returns the distance matrix between LAP trajectories\n",
    "def distanceMatrix2(paths, file_name):\n",
    "    #toReturn = np.zeros((len(paths), len(paths)))\n",
    "    #toReturn = []\n",
    "    file = open(file_name, 'w')\n",
    "    row = 0 \n",
    "    col = 0\n",
    "    while row < len(paths):\n",
    "        toAdd = []\n",
    "        while col < len(paths):\n",
    "            dist, _ = fastdtw(paths[row], paths[col])\n",
    "            toAdd.append(dist)\n",
    "            #file.write(str(dist) + \"\\t\")\n",
    "            #toReturn[row, col] = dist\n",
    "            #toReturn[col, row] = dist\n",
    "            col += 1\n",
    "        file.write(\"\\n\")\n",
    "        col = row+1\n",
    "        row += 1\n",
    "    return toReturn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "06bd263a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPath(predecessors, start_index, end_index):\n",
    "    path = []\n",
    "    while predecessors[start_index]!=-9999:\n",
    "        path.append(start_index)\n",
    "        start_index = predecessors[start_index]\n",
    "    if predecessors[start_index]!=9999:   \n",
    "        path.append(end_index)\n",
    "    return path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "62d6c80d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getDistance(coord1, coord2):\n",
    "    return np.sqrt((coord1[0] - coord2[0])**2 + (coord1[1] - coord2[1])**2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "739bdffd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "0a54084e6b208ee8d1ce3989ffc20924477a5f55f5a43e22e699a6741623861e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
